{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GermlineMasterWorkflow.wdl output check development notes\n",
    "## (Not a cheesy indictment of any known programmer or programming organization)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "x_dir = '/Users/yo/zzIForge/fullyJan10/'\n",
    "if not os.path.isdir(x_dir):\n",
    "    print('directory not found\\n', x_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "****\n",
    "## GermlineMasterWorkflow.wdl call sequence:\n",
    "    * trimseq\n",
    "    * align\n",
    "    * merge\n",
    "    * dedup\n",
    "    * DAB\n",
    "    * realign\n",
    "    * bsqr\n",
    "    * haplotype\n",
    "    * vsqr\n",
    "    * DHVC\n",
    "****\n",
    "****\n",
    "### 1.0) TestTasks/Runtrim_sequences.wdl, workflow as CUTADAPTTRIM.RunTrimSequencesTask (as trimseq)\n",
    "    * passes NormalInputReads as InputReads\n",
    "### 1.5 ) \"trimseq\"  calls Tasks/trim_sequences.wdl workflow as TRIMSEQ.trimsequencesTask as either:\n",
    "#### 1.5.1 ) TRIMSEQ_paired\n",
    "#### 1.5.2 ) TRIMSEQ_single\n",
    "##### >> by either name they (it) runs:\n",
    "    * src/shell/shell_preamble.sh as \"BashPreamble\"\n",
    "    * src/shell/trim_sequences.sh as \"TrimSeqScript\"\n",
    "        * calling src/shell/shared_functions.sh as \"BashSharedFunctions\"\n",
    "##### >> with parameters:\n",
    "    * PairedEnd (Boolean)\n",
    "    * InputRead1 (File)\n",
    "    * InputRead2 (String)\n",
    "    * SampleName (String)\n",
    "    * Adapters (File)\n",
    "    * CutAdapt (String) == path\n",
    "    * CutAdaptThreads (String) == integer\n",
    "    * TrimEnvProfile (File)\n",
    "    * DebugMode (String) == DNE\n",
    "##### >> and orphan parameters:\n",
    "    * CutAdaptThreads aka cpu\n",
    "    * TrimSoftMemLimit aka s_vmem\n",
    "    * TrimHardMemLimit aka h_vmem\n",
    "### 1.5.8 ) Named output\n",
    "    * Outputs Array[File]\n",
    "    * using (unix, linux, ? dakine) \"glob\" function \n",
    "### 1.5.9 ) Output assumption is optional input for \"align\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "****\n",
    "### 2.0 ) TestTasks/Runalignment.wdl, workflow as ALIGNMENT.RunAlignmentTask (as align)\n",
    "### 2.5 ) \"align\" calls Tasks/alignment.wdl, workflow as ALIGN.alignmentTask (as ALIGN_paired)\n",
    "#### 2.5.1 ) ALIGN_paired runs:\n",
    "    * src/shell/shell_preamble.sh as \"BashPreamble\"\n",
    "    * src/shell/alignment.sh as \"AlignmentScript\"\n",
    "        * /usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01 as \"Sentieon\"\n",
    "        * src/shell/shared_functions.sh as \"BashSharedFunctions\"\n",
    "##### >> with parameters:\n",
    "    * PairedEnd (Boolean)\n",
    "    * InputRead1 (File)\n",
    "    * InputRead2 (String)\n",
    "    * SampleName (String)\n",
    "    * Platform (String)\n",
    "    * Library (String)\n",
    "    * PlatformUnit (String)\n",
    "    * CenterName (String)\n",
    "    * Sentieon (String) == path\n",
    "    * SentieonThreads (String) == integer\n",
    "    * AlignEnvProfile (File)\n",
    "    * ChunkSizeInBases (String) \n",
    "    * BWAExtraOptionsString (String)\n",
    "    \n",
    "##### >> and orphan parameters:\n",
    "    * SentieonThreads aka cpu\n",
    "    * AlignSoftMemLimit aka s_vmem\n",
    "    * AlignHardMemLimit aka h_vmem\n",
    "\n",
    "### 2.5.7 ) Named output inner:\n",
    "    * OutputBams: SampleName.aligned.sorted.bam\n",
    "    * OutputBais: SampleName.aligned.sorted.bam.bai\n",
    "### 2.5.8 ) Named output middle:\n",
    "    * OutputBams: Array[File]\n",
    "    * OutputBais: Array[File]\n",
    "### 2.5.9 ) Output assumption is inputs for \"merge\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4) GermlineMasterWorkflow.wdl calls Tasks/merge_aligned_bam.wdl runs:\n",
    "    * MergeBamScript: src/shell/merge_bams.sh (using shared_functions.sh)\n",
    "    * with Config/MergeBamEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.merged.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.merged.bam.bai\n",
    "        \n",
    "5) GermlineMasterWorkflow.wdl calls Tasks/dedup.wdl runs:\n",
    "    * DedupScript: src/shell/dedup.sh (using shared_functions.sh)\n",
    "    * with Config/DedupEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.deduped.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.deduped.bam.bai\n",
    "        \n",
    "6) GermlineMasterWorkflow.wdl calls Tasks/deliver_alignment.wdl runs:\n",
    "    * DeliveryAlignment_Script: src/shell/deliver_alignment.sh (using shared_functions.sh)\n",
    "    * using: GermlineMasterWorkflow.FilledIn.json\n",
    "    * no output but: dowop Delivery/Alignment\n",
    "    \n",
    "7) GermlineMasterWorkflow.wdl calls Tasks/realignment.wdl runs:\n",
    "    * RealignmentScript: src/shell/realignment.sh (using shared_functions.sh)\n",
    "    * using: Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf\n",
    "    * RealignEnvProfile: Config/RealignEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputBams: SampleName.aligned.sorted.deduped.realigned.bam\n",
    "        * OutputBais: SampleName.aligned.sorted.deduped.realigned.bam.bai\n",
    "        \n",
    "8) GermlineMasterWorkflow.wdl calls Tasks/bqsr.wdl runs:\n",
    "    * BqsrScript: src/shell/bqsr.sh (\n",
    "    * using: \n",
    "        * Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf\n",
    "        * Reference/dbsnp_138.hg38.vcf\n",
    "        * Config/BqsrEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * RecalTable: SampleName.recal_data.table\n",
    "        \n",
    "9) GermlineMasterWorkflow.wdl calls Tasks/haplotyper.wdl runs:\n",
    "    * HaplotyperScript: src/shell/haplotyper.sh (using shared_functions.sh)\n",
    "    * using:\n",
    "        * HaplotyperExtraOptionsString: (magic string)\n",
    "        * Config/HaplotyperEnvProfile.file\n",
    "        * Config/HaplotyperVCFSourceField.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputVcf: SampleName.vcf\n",
    "        * OutputVcfIdx: SampleName.vcf.idx\"\n",
    "        \n",
    "10) GermlineMasterWorkflow.wdl calls Tasks/vqsr.wdl runs:\n",
    "    * VqsrScript: src/shell/vqsr.sh \n",
    "    * using magic strings:\n",
    "        * VqsrSnpResourceString\n",
    "        * VqsrIndelResourceString\n",
    "        * AnnotateText\n",
    "    * using:\n",
    "        * InputVcf && InputVcfIdx (input output from) VCF from Haplotyper \n",
    "        * Config/VqsrEnvProfile.file\n",
    "    * Warns: cpu, s_vmem, h_vmem\n",
    "    * Puts Out:\n",
    "        * OutputVcf: SampleName.SNP.recaled.vcf\n",
    "        * OutputVcfIdx: SampleName.SNP.recaled.vcf.idx\n",
    "        \n",
    "10) GermlineMasterWorkflow.wdl calls Tasks/deliver_HaplotyperVC.wdl runs:\n",
    "    * DeliveryHaplotyperVC_Script: src/shell/deliver_haplotyperVC.sh (using shared_functions.sh)\n",
    "    * to dakine: Delivery/HaplotyperVC\n",
    "    * using: GermlineMasterWorkflow.FilledIn.json    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import knpackage.toolbox as kn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 GermlineMasterWF.DAB.WorkflowJson\n",
      "\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "\n",
      "14 GermlineMasterWF.DHVC.WorkflowJson\n",
      "\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "\n"
     ]
    }
   ],
   "source": [
    "json_dir = '../data/Jsons'\n",
    "json_farameters_file = 'GermlineMasterWorkflow.FilledIn.json'\n",
    "\n",
    "json_parameters = kn.get_run_parameters(json_dir, json_farameters_file)\n",
    "dakine = 'WorkflowJson'\n",
    "lineno = 0\n",
    "for k, v in json_parameters.items():\n",
    "    if dakine in k:\n",
    "        print('%i %30s\\n\\t%s\\n'%(lineno, k, v))\n",
    "    lineno += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
