{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GermlineMasterWorkflow.wdl',\n",
       " 'DeliveryOfHaplotyperVC',\n",
       " 'DeliveryOfAlignment',\n",
       " 'SomaticMasterWorkflow.wdl',\n",
       " 'HaplotyperVC',\n",
       " 'Alignment',\n",
       " 'SomaticVC',\n",
       " 'DeliveryOfSomaticVC']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "# pandas future quard\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from collections import OrderedDict\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import yaml\n",
    "\n",
    "clone_base = '/projects/bioinformatics/DEL'\n",
    "config_base = clone_base\n",
    "\n",
    "wdl_directory = os.path.join(clone_base,'MayomicsVC/src/wdl')\n",
    "\n",
    "os.listdir(wdl_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 5 unique types in all the wdl files:\n",
      "\n",
      "\t Array[File]\n",
      "\t Boolean\n",
      "\t File\n",
      "\t File?\n",
      "\t String\n"
     ]
    }
   ],
   "source": [
    "def get_task_types(wdl_directory=None):\n",
    "    \"\"\" Usage: types_list = get_task_types(wdl_directory)\n",
    "    get the unique list of types in the task section of all wdl files in the dir tree \n",
    "    \"\"\"\n",
    "    if not wdl_directory is None and os.path.isdir(wdl_directory):\n",
    "        wdl_directory = wdl_directory\n",
    "    else:\n",
    "        wdl_directory = os.getcwd()\n",
    "\n",
    "    skip_line_chars = ['#', '<', '>']\n",
    "    switch_ON_words = ['task']\n",
    "    switch_OFF_words = ['command', 'runtime', 'output']\n",
    "    add_words_ON = False\n",
    "\n",
    "    first_words = []\n",
    "    for this_dir, dirs, files in os.walk(wdl_directory):\n",
    "        for file in files:\n",
    "            if file[-4:] == '.wdl':\n",
    "                add_words_ON = False\n",
    "                full_filename = os.path.join(this_dir, file)\n",
    "                with open(full_filename, 'r') as fh:\n",
    "                    lines = fh.readlines()\n",
    "                for line in lines:\n",
    "                    l = line.strip()\n",
    "                    if len(l) > 0 and not l[0] in skip_line_chars:\n",
    "                        first_word = l.split()[0]\n",
    "                        if first_word in switch_OFF_words:\n",
    "                            add_words_ON = False\n",
    "                        if add_words_ON == True:\n",
    "                            first_words.append(first_word)\n",
    "                        if first_word in switch_ON_words:\n",
    "                            add_words_ON = True\n",
    "\n",
    "    return sorted(list(set(first_words)))\n",
    "\n",
    "types_list = get_task_types(wdl_directory)\n",
    "print('There are %i unique types in all the wdl files:\\n'%(len(types_list)))\n",
    "for wurd in types_list:\n",
    "    print('\\t',wurd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found 117 variables\n",
      "                      Adapters:               \"File\"\n",
      "               AlignEnvProfile:               \"File\"\n",
      "             AlignHardMemLimit:             \"String\"\n",
      "               AlignOutputBais:        \"Array[File]\"\n",
      "               AlignOutputBams:        \"Array[File]\"\n",
      "             AlignSoftMemLimit:             \"String\"\n",
      "               AlignmentScript:               \"File\"\n",
      "                  AnnotateText:             \"String\"\n",
      "         BWAExtraOptionsString:             \"String\"\n",
      "                  BashPreamble:               \"File\"\n",
      "           BashSharedFunctions:               \"File\"\n",
      "                          Bqsr:            \"Boolean\"\n",
      "                BqsrEnvProfile:               \"File\"\n",
      "              BqsrHardMemLimit:             \"String\"\n",
      "                BqsrKnownSites:             \"String\"\n",
      "                    BqsrScript:               \"File\"\n",
      "              BqsrSoftMemLimit:             \"String\"\n",
      "                    CenterName:             \"String\"\n",
      "              ChunkSizeInBases:             \"String\"\n",
      "                      CutAdapt:             \"String\"\n",
      "               CutAdaptThreads:             \"String\"\n",
      "                         DBSNP:               \"File\"\n",
      "                      DBSNPIdx:               \"File\"\n",
      "                     DebugMode:             \"String\"\n",
      "               DedupEnvProfile:               \"File\"\n",
      "             DedupHardMemLimit:             \"String\"\n",
      "                   DedupScript:               \"File\"\n",
      "             DedupSoftMemLimit:             \"String\"\n",
      "        DeliverAlignOutputBais:               \"File\"\n",
      "        DeliverAlignOutputBams:               \"File\"\n",
      "          DeliverHaplotyperVcf:               \"File\"\n",
      "       DeliverHaplotyperVcfIdx:               \"File\"\n",
      "      DeliveryAlignment_Script:               \"File\"\n",
      "      DeliveryFolder_Alignment:             \"String\"\n",
      "   DeliveryFolder_HaplotyperVC:             \"String\"\n",
      "      DeliveryFolder_SomaticVC:             \"String\"\n",
      "   DeliveryHaplotyperVC_Script:               \"File\"\n",
      "      DeliverySomaticVC_Script:               \"File\"\n",
      " GlobalAlignedSortedDedupedBam:               \"File\"\n",
      "GlobalAlignedSortedDedupedBamBai:               \"File\"\n",
      "         GlobalRecalibratedVcf:               \"File\"\n",
      "      GlobalRecalibratedVcfIdx:               \"File\"\n",
      "          HaplotyperEnvProfile:               \"File\"\n",
      "  HaplotyperExtraOptionsString:             \"String\"\n",
      "        HaplotyperHardMemLimit:             \"String\"\n",
      "              HaplotyperScript:               \"File\"\n",
      "        HaplotyperSoftMemLimit:             \"String\"\n",
      "      HaplotyperVCFSourceField:             \"String\"\n",
      "                     InputBais:  \"File, Array[File]\"\n",
      "                     InputBams:  \"File, Array[File]\"\n",
      "                InputMutectVcf:               \"File\"\n",
      "             InputMutectVcfIdx:               \"File\"\n",
      "                    InputRead1:               \"File\"\n",
      "                    InputRead2:             \"String\"\n",
      "               InputStrelkaVcf:               \"File\"\n",
      "            InputStrelkaVcfIdx:               \"File\"\n",
      "                      InputVcf:               \"File\"\n",
      "                   InputVcfIdx:               \"File\"\n",
      "                       Library:             \"String\"\n",
      "                MarkDuplicates:            \"Boolean\"\n",
      "            MergeBamEnvProfile:               \"File\"\n",
      "                MergeBamScript:               \"File\"\n",
      "             MergeHardMemLimit:             \"String\"\n",
      "             MergeSoftMemLimit:             \"String\"\n",
      "     MergeSomaticVcfEnvProfile:               \"File\"\n",
      "         MergeSomaticVcfScript:               \"File\"\n",
      "                        Mutect:             \"String\"\n",
      "              MutectEnvProfile:               \"File\"\n",
      "      MutectExtraOptionsString:             \"String\"\n",
      "                  MutectScript:               \"File\"\n",
      "                 MutectThreads:             \"String\"\n",
      "                    NormalBais:               \"File\"\n",
      "                    NormalBams:               \"File\"\n",
      "                    OutputBais: \"File, Array[File], File, File, File\"\n",
      "                    OutputBams: \"File, Array[File], File, File, File\"\n",
      "                     OutputVcf:               \"File\"\n",
      "                  OutputVcfIdx:               \"File\"\n",
      "                       Outputs:        \"Array[File]\"\n",
      "                     PairedEnd:            \"Boolean\"\n",
      "                      Platform:             \"String\"\n",
      "                  PlatformUnit:             \"String\"\n",
      "             RealignEnvProfile:               \"File\"\n",
      "           RealignHardMemLimit:             \"String\"\n",
      "           RealignSoftMemLimit:             \"String\"\n",
      "         RealignmentKnownSites:             \"String\"\n",
      "             RealignmentScript:               \"File\"\n",
      "                    RecalTable:        \"File, File?\"\n",
      "                           Ref:               \"File\"\n",
      "                        RefAmb:               \"File\"\n",
      "                        RefAnn:               \"File\"\n",
      "                        RefBwt:               \"File\"\n",
      "                        RefFai:               \"File\"\n",
      "                        RefPac:               \"File\"\n",
      "                         RefSa:               \"File\"\n",
      "                    SampleName:             \"String\"\n",
      "                      Sentieon:             \"String\"\n",
      "               SentieonThreads:             \"String\"\n",
      "                       Strelka:             \"String\"\n",
      "             StrelkaEnvProfile:               \"File\"\n",
      "     StrelkaExtraOptionsString:             \"String\"\n",
      "                 StrelkaScript:               \"File\"\n",
      "                StrelkaThreads:             \"String\"\n",
      "                TrimEnvProfile:               \"File\"\n",
      "              TrimHardMemLimit:             \"String\"\n",
      "                 TrimSeqScript:               \"File\"\n",
      "              TrimSoftMemLimit:             \"String\"\n",
      "                      Trimming:            \"Boolean\"\n",
      "                     TumorBais:               \"File\"\n",
      "                     TumorBams:               \"File\"\n",
      "                          Vqsr:            \"Boolean\"\n",
      "                VqsrEnvProfile:               \"File\"\n",
      "              VqsrHardMemLimit:             \"String\"\n",
      "       VqsrIndelResourceString:             \"String\"\n",
      "                    VqsrScript:               \"File\"\n",
      "         VqsrSnpResourceString:             \"String\"\n",
      "              VqsrSoftMemLimit:             \"String\"\n",
      "                  WorkflowJson:               \"File\"\n"
     ]
    }
   ],
   "source": [
    "def get_wdl_variables_dict(wdl_directory=None):\n",
    "    \"\"\" Usage  config_orderd_dict = get_wdl_variables_dict(wdl_directory)\n",
    "    Get the sorted list of variables and their types from the wdl files in a directory tree\n",
    "    \n",
    "    Args:\n",
    "        wdl_directory:       (default to run directory if not valid directory name)\n",
    "        \n",
    "    Returns:\n",
    "        config_ordered_dict: python OrderedDict of  variable_name: \"type_name\"\n",
    "        \n",
    "    \"\"\"\n",
    "    # Get the complete config.txt file:\n",
    "    skip_line_chars = ['#', '<', '>']\n",
    "    task_types_list = sorted(['Array[File]', 'File', 'File?', 'Boolean', 'String'])\n",
    "    config_vars_dict = {}\n",
    "    for this_dir, dirs, files in os.walk(wdl_directory):\n",
    "        for file in files:\n",
    "            if file[-4:] == '.wdl' and file[0] != '.':\n",
    "                full_filename = os.path.join(this_dir, file)\n",
    "                with open(full_filename, 'r') as fh:\n",
    "                    lines = fh.readlines()\n",
    "                for line in lines:\n",
    "                    l = line.strip()\n",
    "                    if len(l) > 0 and not l[0] in skip_line_chars:\n",
    "                        line_words_list = l.split()\n",
    "                        first_word = line_words_list[0]\n",
    "                        if first_word in task_types_list:\n",
    "                            second_word = line_words_list[1]\n",
    "                            if not second_word in config_vars_dict.keys():\n",
    "                                config_vars_dict[second_word] = '\"' + first_word + '\"'\n",
    "                                \n",
    "                            elif first_word != config_vars_dict[second_word][1:-1]:\n",
    "                                bugger = config_vars_dict[second_word][:-1]\n",
    "                                config_vars_dict[second_word] = bugger + ', ' + first_word  + '\"'\n",
    "                                \n",
    "    config_od = OrderedDict()\n",
    "    for k, v in sorted(config_vars_dict.items()):\n",
    "        config_od[k] = v\n",
    "\n",
    "    return config_od\n",
    "\n",
    "config_vars_dict_ret = get_wdl_variables_dict(wdl_directory)\n",
    "print('found %i variables'%(len(config_vars_dict_ret)))\n",
    "config_od = OrderedDict()\n",
    "for k, v in sorted(config_vars_dict_ret.items()):\n",
    "    config_od[k] = v\n",
    "    \n",
    "for k, v in config_od.items():\n",
    "    print('%30s: %20s'%(k,v))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(30, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_path</th>\n",
       "      <th>imports</th>\n",
       "      <th>workflows</th>\n",
       "      <th>tasks</th>\n",
       "      <th>inputs</th>\n",
       "      <th>outputs</th>\n",
       "      <th>commands</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDL file</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWorkflow.wdl</th>\n",
       "      <td>../wdl</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SomaticMasterWorkflow.wdl</th>\n",
       "      <td>../wdl</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DeliverHaplotyperVC_ForWorkflow.wdl</th>\n",
       "      <td>../wdl/DeliveryOfHaplotyperVC/Workflow</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdeliver_haplotyperVC.wdl</th>\n",
       "      <td>../wdl/DeliveryOfHaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deliver_HaplotyperVC.wdl</th>\n",
       "      <td>../wdl/DeliveryOfHaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TestDeliverAlignment.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/Workflow</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DeliverAlignment_ForWorkflow.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/Workflow</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdeliver_alignment.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deliver_alignment.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TestHaplotyperVC.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Workflow</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testhaplotyper.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testvqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testbqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testrealignment.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>haplotyper.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>realignment.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TestAlignment.wdl</th>\n",
       "      <td>../wdl/Alignment/Workflow</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Runalignment.wdl</th>\n",
       "      <td>../wdl/Alignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdedup.wdl</th>\n",
       "      <td>../wdl/Alignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Runtrim_sequences.wdl</th>\n",
       "      <td>../wdl/Alignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alignment.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trim_sequences.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dedup.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>merge_aligned_bam.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strelka.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mutect.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>merge_somatic_vcf.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deliver_SomaticVC.wdl</th>\n",
       "      <td>../wdl/DeliveryOfSomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                    src_path  \\\n",
       "WDL file                                                                       \n",
       "GermlineMasterWorkflow.wdl                                            ../wdl   \n",
       "SomaticMasterWorkflow.wdl                                             ../wdl   \n",
       "DeliverHaplotyperVC_ForWorkflow.wdl   ../wdl/DeliveryOfHaplotyperVC/Workflow   \n",
       "Testdeliver_haplotyperVC.wdl         ../wdl/DeliveryOfHaplotyperVC/TestTasks   \n",
       "deliver_HaplotyperVC.wdl                 ../wdl/DeliveryOfHaplotyperVC/Tasks   \n",
       "TestDeliverAlignment.wdl                 ../wdl/DeliveryOfAlignment/Workflow   \n",
       "DeliverAlignment_ForWorkflow.wdl         ../wdl/DeliveryOfAlignment/Workflow   \n",
       "Testdeliver_alignment.wdl               ../wdl/DeliveryOfAlignment/TestTasks   \n",
       "deliver_alignment.wdl                       ../wdl/DeliveryOfAlignment/Tasks   \n",
       "TestHaplotyperVC.wdl                            ../wdl/HaplotyperVC/Workflow   \n",
       "Testhaplotyper.wdl                             ../wdl/HaplotyperVC/TestTasks   \n",
       "Testvqsr.wdl                                   ../wdl/HaplotyperVC/TestTasks   \n",
       "Testbqsr.wdl                                   ../wdl/HaplotyperVC/TestTasks   \n",
       "Testrealignment.wdl                            ../wdl/HaplotyperVC/TestTasks   \n",
       "bqsr.wdl                                           ../wdl/HaplotyperVC/Tasks   \n",
       "haplotyper.wdl                                     ../wdl/HaplotyperVC/Tasks   \n",
       "vqsr.wdl                                           ../wdl/HaplotyperVC/Tasks   \n",
       "realignment.wdl                                    ../wdl/HaplotyperVC/Tasks   \n",
       "TestAlignment.wdl                                  ../wdl/Alignment/Workflow   \n",
       "Runalignment.wdl                                  ../wdl/Alignment/TestTasks   \n",
       "Testdedup.wdl                                     ../wdl/Alignment/TestTasks   \n",
       "Runtrim_sequences.wdl                             ../wdl/Alignment/TestTasks   \n",
       "alignment.wdl                                         ../wdl/Alignment/Tasks   \n",
       "trim_sequences.wdl                                    ../wdl/Alignment/Tasks   \n",
       "dedup.wdl                                             ../wdl/Alignment/Tasks   \n",
       "merge_aligned_bam.wdl                                 ../wdl/Alignment/Tasks   \n",
       "strelka.wdl                                           ../wdl/SomaticVC/Tasks   \n",
       "mutect.wdl                                            ../wdl/SomaticVC/Tasks   \n",
       "merge_somatic_vcf.wdl                                 ../wdl/SomaticVC/Tasks   \n",
       "deliver_SomaticVC.wdl                       ../wdl/DeliveryOfSomaticVC/Tasks   \n",
       "\n",
       "                                     imports  workflows  tasks  inputs  \\\n",
       "WDL file                                                                 \n",
       "GermlineMasterWorkflow.wdl                10          1      0      10   \n",
       "SomaticMasterWorkflow.wdl                 10          4      0      16   \n",
       "DeliverHaplotyperVC_ForWorkflow.wdl        1          1      0       1   \n",
       "Testdeliver_haplotyperVC.wdl               1          1      0       0   \n",
       "deliver_HaplotyperVC.wdl                   0          1      1       0   \n",
       "TestDeliverAlignment.wdl                   1          1      0       0   \n",
       "DeliverAlignment_ForWorkflow.wdl           1          1      0       1   \n",
       "Testdeliver_alignment.wdl                  1          1      0       0   \n",
       "deliver_alignment.wdl                      0          1      1       0   \n",
       "TestHaplotyperVC.wdl                       4          1      0       3   \n",
       "Testhaplotyper.wdl                         1          1      0       0   \n",
       "Testvqsr.wdl                               1          1      0       0   \n",
       "Testbqsr.wdl                               1          1      0       0   \n",
       "Testrealignment.wdl                        1          1      0       0   \n",
       "bqsr.wdl                                   0          0      2       1   \n",
       "haplotyper.wdl                             0          0      2       0   \n",
       "vqsr.wdl                                   0          0      2       1   \n",
       "realignment.wdl                            0          0      1       0   \n",
       "TestAlignment.wdl                          4          1      0       3   \n",
       "Runalignment.wdl                           1          1      1       3   \n",
       "Testdedup.wdl                              1          1      0       0   \n",
       "Runtrim_sequences.wdl                      1          1      0       3   \n",
       "alignment.wdl                              0          0      1       0   \n",
       "trim_sequences.wdl                         0          0      1       0   \n",
       "dedup.wdl                                  0          0      2       1   \n",
       "merge_aligned_bam.wdl                      0          0      1       1   \n",
       "strelka.wdl                                0          0      1       0   \n",
       "mutect.wdl                                 0          0      1       0   \n",
       "merge_somatic_vcf.wdl                      0          0      2       0   \n",
       "deliver_SomaticVC.wdl                      0          1      1       0   \n",
       "\n",
       "                                     outputs  commands  \n",
       "WDL file                                                \n",
       "GermlineMasterWorkflow.wdl                 0         0  \n",
       "SomaticMasterWorkflow.wdl                  0         0  \n",
       "DeliverHaplotyperVC_ForWorkflow.wdl        0         0  \n",
       "Testdeliver_haplotyperVC.wdl               0         0  \n",
       "deliver_HaplotyperVC.wdl                   0         1  \n",
       "TestDeliverAlignment.wdl                   0         0  \n",
       "DeliverAlignment_ForWorkflow.wdl           0         0  \n",
       "Testdeliver_alignment.wdl                  0         0  \n",
       "deliver_alignment.wdl                      1         1  \n",
       "TestHaplotyperVC.wdl                       0         0  \n",
       "Testhaplotyper.wdl                         0         0  \n",
       "Testvqsr.wdl                               0         0  \n",
       "Testbqsr.wdl                               0         0  \n",
       "Testrealignment.wdl                        0         0  \n",
       "bqsr.wdl                                   1         1  \n",
       "haplotyper.wdl                             1         1  \n",
       "vqsr.wdl                                   1         1  \n",
       "realignment.wdl                            1         1  \n",
       "TestAlignment.wdl                          1         0  \n",
       "Runalignment.wdl                           2         0  \n",
       "Testdedup.wdl                              0         0  \n",
       "Runtrim_sequences.wdl                      2         0  \n",
       "alignment.wdl                              1         1  \n",
       "trim_sequences.wdl                         1         1  \n",
       "dedup.wdl                                  1         1  \n",
       "merge_aligned_bam.wdl                      1         1  \n",
       "strelka.wdl                                1         1  \n",
       "mutect.wdl                                 1         1  \n",
       "merge_somatic_vcf.wdl                      1         1  \n",
       "deliver_SomaticVC.wdl                      0         1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_wdl_files_spreadsheet(wdl_directory=None):\n",
    "    \"\"\" wdl_files_spreadsheet = get_wdl_files_spreadsheet(wdl_directory)\n",
    "    get the section statistics spreadsheet for a directory tree of wdl files \n",
    "    \n",
    "    Args:\n",
    "        wdl_directory:       (default to run directory if not valid directory name)\n",
    "        \n",
    "    Returns:\n",
    "        wdl_files_spreadsheet: pandas dataframe with rows of wdl filenames, cols of wdl keywords\n",
    "\n",
    "    \"\"\"\n",
    "    if not wdl_directory is None and os.path.isdir(wdl_directory):\n",
    "        may_dir = wdl_directory\n",
    "    else:\n",
    "        may_dir = os.getcwd()\n",
    "    \n",
    "    dir_name_stripper, _ = os.path.split(wdl_directory)\n",
    "\n",
    "    prefix_ignores = ['.', '_']\n",
    "    \n",
    "    # assemble dictionary - wdl file names: full path  -------- Build empty pandas dataframe\n",
    "    mdl_dict = {}\n",
    "    for dir_name, dir_list, files_list in os.walk(may_dir):\n",
    "        if len(files_list) > 0:\n",
    "            for file_name in files_list:\n",
    "                if file_name[-4:] == '.wdl' and file_name[0] not in prefix_ignores:\n",
    "                    mdl_dict[file_name] = os.path.join(dir_name, file_name)\n",
    "    wdl_df = pd.DataFrame.from_dict(mdl_dict, orient='index', columns=['src_path'])\n",
    "    wdl_df.index.name = 'WDL file'\n",
    "    \n",
    "    # define the rest of the columns in terms of wdl keywords\n",
    "    srch_dict = OrderedDict([('imports','import'), \n",
    "                             ('workflows','workflow'), \n",
    "                            ('tasks','task'), \n",
    "                            ('inputs','input'), \n",
    "                            ('outputs','output'),\n",
    "                            ('commands', 'command')])\n",
    "    for col_name, _ in srch_dict.items():\n",
    "        wdl_df[col_name] = 0\n",
    "\n",
    "    # for each wdl file: count keywords, get top path ---------- fill pandas dataframe\n",
    "    for name, row in wdl_df.iterrows():\n",
    "        \"\"\" using row.src_path to open files and name to refer to row in dataframe \"\"\"\n",
    "        lines = []\n",
    "        try:\n",
    "            with open(row.src_path, 'r') as fh:\n",
    "                lines = fh.readlines()\n",
    "        except:\n",
    "            print('Fails to open:\\n', row.src_path)\n",
    "            pass\n",
    "        \n",
    "        # count & insert occurrence of keywords\n",
    "        if len(lines) > 0:\n",
    "            for l in lines:\n",
    "                for col_name, key_word in srch_dict.items():\n",
    "                    if key_word in l:\n",
    "                        wdl_df[col_name].loc[name] += 1\n",
    "\n",
    "        # replace the full path name with top path name for readability\n",
    "        full_path, _ = os.path.split(mdl_dict[name])\n",
    "        wdl_df['src_path'].loc[name] = full_path.replace(dir_name_stripper, '..')\n",
    "        \n",
    "    return wdl_df\n",
    "\n",
    "wdl_df = get_wdl_files_spreadsheet(wdl_directory)\n",
    "print(wdl_df.shape)\n",
    "wdl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spreadsheet_size =  (193, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variable Value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variable Name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.Bqsr</th>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.BashPreamble</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.BashSharedFunctions</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.DebugMode</th>\n",
       "      <td>-d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.DeliveryAlignment_Script</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.DeliveryFolder_Alignment</th>\n",
       "      <td>/projects/bioinformatics/DEL/Delivery/Alignmen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.SampleName</th>\n",
       "      <td>WGS_chr20_21_22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DAB.WorkflowJson</th>\n",
       "      <td>/projects/bioinformatics/DEL/Jsons/SomaticMast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.BashPreamble</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.BashSharedFunctions</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.DebugMode</th>\n",
       "      <td>-d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC</th>\n",
       "      <td>/projects/bioinformatics/DEL/Delivery/Haplotyp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.SampleName</th>\n",
       "      <td>WGS_chr20_21_22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.DHVC.WorkflowJson</th>\n",
       "      <td>/projects/bioinformatics/DEL/Jsons/SomaticMast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.MarkDuplicates</th>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.NormalInputReads</th>\n",
       "      <td>[[/projects/bioinformatics/MRW/MayomicsVC-Dev/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.Trimming</th>\n",
       "      <td>false</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.Vqsr</th>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.AlignEnvProfile</th>\n",
       "      <td>/projects/bioinformatics/DEL/Config/AlignEnvPr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit</th>\n",
       "      <td>100G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit</th>\n",
       "      <td>100G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.AlignmentScript</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString</th>\n",
       "      <td>\"'-M'\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.BashPreamble</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.BashSharedFunctions</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.CenterName</th>\n",
       "      <td>some_seq_center</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases</th>\n",
       "      <td>10000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.DebugMode</th>\n",
       "      <td>-d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.align.ALIGN_paired.Library</th>\n",
       "      <td>fake_lib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt</th>\n",
       "      <td>/usr/local/apps/bioapps/python/Python-3.6.1/bin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode</th>\n",
       "      <td>-d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd</th>\n",
       "      <td>true</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.SampleName</th>\n",
       "      <td>WGS_chr20_21_22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile</th>\n",
       "      <td>/projects/bioinformatics/DEL/Config/TrimEnvPro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit</th>\n",
       "      <td>100G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit</th>\n",
       "      <td>100G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.AnnotateText</th>\n",
       "      <td>\"'--annotation DP --annotation QD --annotation...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.BashPreamble</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.BashSharedFunctions</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.DebugMode</th>\n",
       "      <td>-d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.Ref</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefAmb</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefAnn</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefBwt</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefDict</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefFai</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefPac</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.RefSa</th>\n",
       "      <td>/projects/mgc/Project_1/MayomicsVC_MayoTesting...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.SampleName</th>\n",
       "      <td>WGS_chr20_21_22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.Sentieon</th>\n",
       "      <td>/usr/local/apps/bioapps/sentieon/sentieon-geno...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.SentieonThreads</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.VqsrEnvProfile</th>\n",
       "      <td>/projects/bioinformatics/DEL/Config/VqsrEnvPro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.VqsrHardMemLimit</th>\n",
       "      <td>100G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.VqsrIndelResourceString</th>\n",
       "      <td>\"'--resource /projects/mgc/Project_1/MayomicsV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.VqsrScript</th>\n",
       "      <td>/projects/bioinformatics/DEL/MayomicsVC/src/sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.VqsrSnpResourceString</th>\n",
       "      <td>\"'--resource /projects/bioinformatics/DataPack...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWF.vqsr.VqsrSoftMemLimit</th>\n",
       "      <td>100G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                       Variable Value\n",
       "Variable Name                                                                                        \n",
       "GermlineMasterWF.Bqsr                                                                            true\n",
       "GermlineMasterWF.DAB.BashPreamble                   /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.DAB.BashSharedFunctions            /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.DAB.DebugMode                                                                     -d\n",
       "GermlineMasterWF.DAB.DeliveryAlignment_Script       /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.DAB.DeliveryFolder_Alignment       /projects/bioinformatics/DEL/Delivery/Alignmen...\n",
       "GermlineMasterWF.DAB.SampleName                                                       WGS_chr20_21_22\n",
       "GermlineMasterWF.DAB.WorkflowJson                   /projects/bioinformatics/DEL/Jsons/SomaticMast...\n",
       "GermlineMasterWF.DHVC.BashPreamble                  /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.DHVC.BashSharedFunctions           /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.DHVC.DebugMode                                                                    -d\n",
       "GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC   /projects/bioinformatics/DEL/Delivery/Haplotyp...\n",
       "GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script   /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.DHVC.SampleName                                                      WGS_chr20_21_22\n",
       "GermlineMasterWF.DHVC.WorkflowJson                  /projects/bioinformatics/DEL/Jsons/SomaticMast...\n",
       "GermlineMasterWF.MarkDuplicates                                                                  true\n",
       "GermlineMasterWF.NormalInputReads                   [[/projects/bioinformatics/MRW/MayomicsVC-Dev/...\n",
       "GermlineMasterWF.Trimming                                                                       false\n",
       "GermlineMasterWF.Vqsr                                                                            true\n",
       "GermlineMasterWF.align.ALIGN_paired.AlignEnvPro...  /projects/bioinformatics/DEL/Config/AlignEnvPr...\n",
       "GermlineMasterWF.align.ALIGN_paired.AlignHardMe...                                               100G\n",
       "GermlineMasterWF.align.ALIGN_paired.AlignSoftMe...                                               100G\n",
       "GermlineMasterWF.align.ALIGN_paired.AlignmentSc...  /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.align.ALIGN_paired.BWAExtraOpt...                                             \"'-M'\"\n",
       "GermlineMasterWF.align.ALIGN_paired.BashPreamble    /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.align.ALIGN_paired.BashSharedF...  /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.align.ALIGN_paired.CenterName                                        some_seq_center\n",
       "GermlineMasterWF.align.ALIGN_paired.ChunkSizeIn...                                           10000000\n",
       "GermlineMasterWF.align.ALIGN_paired.DebugMode                                                      -d\n",
       "GermlineMasterWF.align.ALIGN_paired.Library                                                  fake_lib\n",
       "...                                                                                               ...\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt      /usr/local/apps/bioapps/python/Python-3.6.1/bin\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdap...                                                  8\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode                                                  -d\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd                                                true\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.SampleName                                    WGS_chr20_21_22\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnv...  /projects/bioinformatics/DEL/Config/TrimEnvPro...\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHar...                                               100G\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeq...  /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSof...                                               100G\n",
       "GermlineMasterWF.vqsr.AnnotateText                  \"'--annotation DP --annotation QD --annotation...\n",
       "GermlineMasterWF.vqsr.BashPreamble                  /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.vqsr.BashSharedFunctions           /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.vqsr.DebugMode                                                                    -d\n",
       "GermlineMasterWF.vqsr.Ref                           /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefAmb                        /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefAnn                        /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefBwt                        /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefDict                       /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefFai                        /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefPac                        /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.RefSa                         /projects/mgc/Project_1/MayomicsVC_MayoTesting...\n",
       "GermlineMasterWF.vqsr.SampleName                                                      WGS_chr20_21_22\n",
       "GermlineMasterWF.vqsr.Sentieon                      /usr/local/apps/bioapps/sentieon/sentieon-geno...\n",
       "GermlineMasterWF.vqsr.SentieonThreads                                                              40\n",
       "GermlineMasterWF.vqsr.VqsrEnvProfile                /projects/bioinformatics/DEL/Config/VqsrEnvPro...\n",
       "GermlineMasterWF.vqsr.VqsrHardMemLimit                                                           100G\n",
       "GermlineMasterWF.vqsr.VqsrIndelResourceString       \"'--resource /projects/mgc/Project_1/MayomicsV...\n",
       "GermlineMasterWF.vqsr.VqsrScript                    /projects/bioinformatics/DEL/MayomicsVC/src/sh...\n",
       "GermlineMasterWF.vqsr.VqsrSnpResourceString         \"'--resource /projects/bioinformatics/DataPack...\n",
       "GermlineMasterWF.vqsr.VqsrSoftMemLimit                                                           100G\n",
       "\n",
       "[193 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import yaml\n",
    "\n",
    "def read_json_to_df(full_filename):\n",
    "    \"\"\" Usage: json_df = read_json_to_df(full_filename)\n",
    "    read a json file into a spreadsheet\n",
    "    \"\"\"\n",
    "    if os.path.isfile(full_filename) != True:\n",
    "        run_parameters = {}\n",
    "    else:\n",
    "        with open(full_filename, 'r') as file_handle:\n",
    "            run_parameters = yaml.load(file_handle)\n",
    "        if not isinstance(run_parameters, dict):\n",
    "            run_parameters = {}\n",
    "            \n",
    "    json_df = pd.DataFrame.from_dict(run_parameters, orient='index', columns=['Variable Value'])\n",
    "    json_df.index.name = 'Variable Name'\n",
    "    \n",
    "    return json_df\n",
    "\n",
    "config_base = clone_base\n",
    "try_dir = os.path.join(config_base, 'Jsons/GermlineMasterWorkflow.FilledIn.json')\n",
    "jojo_df = read_json_to_df(try_dir)\n",
    "print('spreadsheet_size = ', jojo_df.shape)\n",
    "jojo_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## develop test script"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "%%writefile check_test_ouput.py\n",
    "\"\"\"\n",
    "check_test_ouput.py\n",
    "Usage:\n",
    "python ~/python/check_test_ouput.py -d /projects/mgc/Project_1/DEL/MVP/cromwell-executions/GermlineMasterWF/\n",
    "or \n",
    "python ~/python/check_test_ouput.py -d `pwd`\n",
    "\n",
    "check return codes and std_out in directory tree\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "good_return_codes_list = ['0', '0\\n']\n",
    "file_names_to_show_list = ['rc', 'stdout', 'stderr']\n",
    "\n",
    "def check_rc_codes(x_directory=None, show_stds=False):\n",
    "    if not x_directory is None and os.path.isdir(x_directory):\n",
    "        dir_tree_root = x_directory\n",
    "    else:\n",
    "        dir_tree_root = os.getcwd()\n",
    "    root_trim_str, _ = os.path.split(dir_tree_root)\n",
    "    for dir_name, dir_list, files_list in os.walk(dir_tree_root):\n",
    "        for filename in files_list:\n",
    "            if filename in file_names_to_show_list:\n",
    "                full_filename = os.path.join(dir_name, filename)\n",
    "                with open(full_filename, 'r') as fh:\n",
    "                    lines = fh.readlines()\n",
    "                if len(lines) <= 0:\n",
    "                    continue\n",
    "                elif filename == 'rc':\n",
    "                    if lines[0] in good_return_codes_list:\n",
    "                        top_dir = dir_name.replace(root_trim_str, '..')\n",
    "                        print('good rc:  %s'%(top_dir))\n",
    "                    else:\n",
    "                        print('\\n\\tBad Dog! Bad Dog!')\n",
    "                        print('code = %s in \\n%s\\n'%(str(lines[0]).strip(), full_filename))\n",
    "                elif show_stds:\n",
    "                    # opportunity to reformat for readability\n",
    "                    for line in lines:\n",
    "                        print(line)\n",
    "                    \n",
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('-d', type=str)\n",
    "    parser.add_argument('-show_stds', nargs=\"?\", const=False, type=bool)\n",
    "    args = parser.parse_args()\n",
    "    check_rc_codes(args.d, args.show_stds)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
