{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Operate the pipetools on iforge head node:\n",
    "\n",
    "## import modules, set warning level, establish directory locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SomaticVC',\n",
       " 'DeliveryOfSomaticVC',\n",
       " 'GermlineMasterWorkflow.wdl',\n",
       " 'Alignment',\n",
       " 'SomaticMasterWorkflow.wdl',\n",
       " 'SomaticVCStandaloneWorkflow.wdl',\n",
       " 'DeliveryOfHaplotyperVC',\n",
       " 'DeliveryOfAlignment',\n",
       " 'HaplotyperVC']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "# pandas future quard\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from collections import OrderedDict\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import yaml\n",
    "\n",
    "# clone_base = '/projects/bioinformatics/DEL'\n",
    "clone_base = '/Users/mojo/git_clone/Mayomics-Research'\n",
    "config_base = clone_base\n",
    "\n",
    "wdl_directory = os.path.join(clone_base,'MayomicsVC/src/wdl')\n",
    "\n",
    "os.listdir(wdl_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse the \"wdl_directory\", read all \".wdl\" files, get types used in the \"task\" sections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 5 types in all the wdl files:\n",
      "\n",
      "                File = 123\n",
      "              String = 114\n",
      "             Boolean = 2\n",
      "         Array[File] = 2\n",
      "               File? = 1\n"
     ]
    }
   ],
   "source": [
    "def get_task_types(wdl_directory=None):\n",
    "    \"\"\" Usage: types_list = get_task_types(wdl_directory)\n",
    "    get the unique list of types in the task section of all wdl files in the dir tree \n",
    "    \"\"\"\n",
    "    if not wdl_directory is None and os.path.isdir(wdl_directory):\n",
    "        wdl_directory = wdl_directory\n",
    "    else:\n",
    "        wdl_directory = os.getcwd()\n",
    "\n",
    "    skip_line_chars = ['#', '<', '>']\n",
    "    switch_ON_words = ['task']\n",
    "    switch_OFF_words = ['command', 'runtime', 'output']\n",
    "    add_words_ON = False\n",
    "\n",
    "    type_words = {}\n",
    "    for this_dir, dirs, files in os.walk(wdl_directory):\n",
    "        for file in files:\n",
    "            if file[-4:] == '.wdl':\n",
    "                add_words_ON = False\n",
    "                full_filename = os.path.join(this_dir, file)\n",
    "                \n",
    "                with open(full_filename, 'r') as fh:\n",
    "                    lines = fh.readlines()\n",
    "                    \n",
    "                for line in lines:\n",
    "                    l = line.strip()\n",
    "                    if len(l) > 0 and not l[0] in skip_line_chars:\n",
    "                        type_word = l.split()[0]\n",
    "                        \n",
    "                        if type_word in switch_OFF_words:\n",
    "                            add_words_ON = False\n",
    "                            \n",
    "                        if add_words_ON == True:\n",
    "                            if type_word in type_words:\n",
    "                                type_words[type_word] += 1\n",
    "                            else:\n",
    "                                type_words[type_word] = 1\n",
    "                            \n",
    "                        if type_word in switch_ON_words:\n",
    "                            add_words_ON = True\n",
    "\n",
    "    return type_words\n",
    "\n",
    "types_dict = get_task_types(wdl_directory)\n",
    "print('There are %i types in all the wdl files:\\n'%(len(types_dict)))\n",
    "for wurd, wurd_count in types_dict.items():\n",
    "    print('%20s = %i'%(wurd, wurd_count))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Walk the 'wdl' directory, parse all wdl files, build a dictionary -- variable names: type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found 158 variables\n",
      "                      Adapters:               \"File\"\n",
      "               AlignEnvProfile:               \"File\"\n",
      "             AlignHardMemLimit:             \"String\"\n",
      "               AlignOutputBais:        \"Array[File]\"\n",
      "               AlignOutputBams:        \"Array[File]\"\n",
      "             AlignSoftMemLimit:             \"String\"\n",
      "               AlignmentScript:               \"File\"\n",
      "                  AnnotateText:             \"String\"\n",
      "         BWAExtraOptionsString:             \"String\"\n",
      "                  BashPreamble:               \"File\"\n",
      "           BashSharedFunctions:               \"File\"\n",
      "                      Bcftools:             \"String\"\n",
      "    BcftoolsExtraOptionsString:             \"String\"\n",
      "                         Bgzip:             \"String\"\n",
      "                          Bqsr:            \"Boolean\"\n",
      "                BqsrEnvProfile:               \"File\"\n",
      "              BqsrHardMemLimit:             \"String\"\n",
      "                BqsrKnownSites:             \"String\"\n",
      "                    BqsrScript:               \"File\"\n",
      "              BqsrSoftMemLimit:             \"String\"\n",
      "                    CenterName:             \"String\"\n",
      "              ChunkSizeInBases:             \"String\"\n",
      "     CombineVariantsEnvProfile:               \"File\"\n",
      "CombineVariantsExtraOptionsString:             \"String\"\n",
      "   CombineVariantsHardMemLimit:             \"String\"\n",
      "         CombineVariantsScript:               \"File\"\n",
      "   CombineVariantsSoftMemLimit:             \"String\"\n",
      "        CombineVariantsThreads:             \"String\"\n",
      "                      CutAdapt:             \"String\"\n",
      "               CutAdaptThreads:             \"String\"\n",
      "                         DBSNP:               \"File\"\n",
      "                      DBSNPIdx:               \"File\"\n",
      "                     DebugMode:             \"String\"\n",
      "               DedupEnvProfile:               \"File\"\n",
      "             DedupHardMemLimit:             \"String\"\n",
      "                   DedupScript:               \"File\"\n",
      "             DedupSoftMemLimit:             \"String\"\n",
      "        DeliverAlignOutputBais:               \"File\"\n",
      "        DeliverAlignOutputBams:               \"File\"\n",
      "          DeliverHaplotyperVcf:               \"File\"\n",
      "       DeliverHaplotyperVcfIdx:               \"File\"\n",
      "      DeliveryAlignment_Script:               \"File\"\n",
      "      DeliveryFolder_Alignment:             \"String\"\n",
      "   DeliveryFolder_HaplotyperVC:             \"String\"\n",
      "      DeliveryFolder_SomaticVC:             \"String\"\n",
      "   DeliveryHaplotyperVC_Script:               \"File\"\n",
      "      DeliverySomaticVC_Script:               \"File\"\n",
      "                   FixDPScript:               \"File\"\n",
      "                       GatkJar:             \"String\"\n",
      " GlobalAlignedSortedDedupedBam:               \"File\"\n",
      "GlobalAlignedSortedDedupedBamBai:               \"File\"\n",
      "         GlobalRecalibratedVcf:               \"File\"\n",
      "      GlobalRecalibratedVcfIdx:               \"File\"\n",
      "          HaplotyperEnvProfile:               \"File\"\n",
      "  HaplotyperExtraOptionsString:             \"String\"\n",
      "        HaplotyperHardMemLimit:             \"String\"\n",
      "              HaplotyperScript:               \"File\"\n",
      "        HaplotyperSoftMemLimit:             \"String\"\n",
      "      HaplotyperVCFSourceField:             \"String\"\n",
      "          IndelGtFixPerlScript:               \"File\"\n",
      "                     InputBais: \"File, Array[File], File, File, File, File\"\n",
      "                     InputBams: \"File, Array[File], File, File, File, File\"\n",
      "                    InputRead1:               \"File\"\n",
      "                    InputRead2:             \"String\"\n",
      "                      InputVcf:               \"File\"\n",
      "                    InputVcfGz:               \"File\"\n",
      "                 InputVcfGzTbi:               \"File\"\n",
      "                   InputVcfIdx:               \"File\"\n",
      "                          Java:             \"String\"\n",
      "                       Library:             \"String\"\n",
      "                         Manta:             \"String\"\n",
      "             MantaAnalysisPath:             \"String\"\n",
      "             MantaHardMemLimit:             \"String\"\n",
      "                   MantaScript:               \"File\"\n",
      "             MantaSoftMemLimit:             \"String\"\n",
      "                  MantaThreads:             \"String\"\n",
      "                MarkDuplicates:            \"Boolean\"\n",
      "            MergeBamEnvProfile:               \"File\"\n",
      "                MergeBamScript:               \"File\"\n",
      "             MergeHardMemLimit:             \"String\"\n",
      "             MergeSoftMemLimit:             \"String\"\n",
      "              MutectEnvProfile:               \"File\"\n",
      "      MutectExtraOptionsString:             \"String\"\n",
      "            MutectHardMemLimit:             \"String\"\n",
      "           MutectJavaMemOption:             \"String\"\n",
      "                  MutectScript:               \"File\"\n",
      "            MutectSoftMemLimit:             \"String\"\n",
      "                 MutectThreads:             \"String\"\n",
      "                  MutectVcfBgz:               \"File\"\n",
      "               MutectVcfBgzTbi:               \"File\"\n",
      "                    NormalBais:               \"File\"\n",
      "                    NormalBams:               \"File\"\n",
      "                    OutputBais: \"File, Array[File], File, File\"\n",
      "                    OutputBams: \"File, Array[File], File, File\"\n",
      "        OutputCandidateSVVcfGz:               \"File\"\n",
      "     OutputCandidateSVVcfGzTbi:               \"File\"\n",
      "OutputCandidateSmallIndelsVcfGz:               \"File\"\n",
      "OutputCandidateSmallIndelsVcfGzTbi:               \"File\"\n",
      "          OutputDiploidSVVcfGz:               \"File\"\n",
      "       OutputDiploidSVVcfGzTbi:               \"File\"\n",
      "                     OutputVcf:               \"File\"\n",
      "                  OutputVcfBgz:               \"File\"\n",
      "               OutputVcfBgzTbi:               \"File\"\n",
      "                   OutputVcfGz:               \"File\"\n",
      "                OutputVcfGzTbi:               \"File\"\n",
      "                  OutputVcfIdx:               \"File\"\n",
      "                       Outputs:        \"Array[File]\"\n",
      "          OutputsomaticSVVcfGz:               \"File\"\n",
      "       OutputsomaticSVVcfGzTbi:               \"File\"\n",
      "                     PairedEnd:            \"Boolean\"\n",
      "                      Platform:             \"String\"\n",
      "                  PlatformUnit:             \"String\"\n",
      "    PrioritizationOptionString:             \"String\"\n",
      "             RealignEnvProfile:               \"File\"\n",
      "           RealignHardMemLimit:             \"String\"\n",
      "           RealignSoftMemLimit:             \"String\"\n",
      "         RealignmentKnownSites:             \"String\"\n",
      "             RealignmentScript:               \"File\"\n",
      "                    RecalTable:        \"File?, File\"\n",
      "                           Ref:               \"File\"\n",
      "                        RefAmb:               \"File\"\n",
      "                        RefAnn:               \"File\"\n",
      "                        RefBwt:               \"File\"\n",
      "                       RefDict:               \"File\"\n",
      "                        RefFai:               \"File\"\n",
      "                        RefPac:               \"File\"\n",
      "                         RefSa:               \"File\"\n",
      "                    SampleName:             \"String\"\n",
      "              SampleNameNormal:             \"String\"\n",
      "               SampleNameTumor:             \"String\"\n",
      "                      Samtools:             \"String\"\n",
      "                      Sentieon:             \"String\"\n",
      "               SentieonThreads:             \"String\"\n",
      "            SnpGtFixPerlScript:               \"File\"\n",
      "                       Strelka:             \"String\"\n",
      "             StrelkaEnvProfile:               \"File\"\n",
      "     StrelkaExtraOptionsString:             \"String\"\n",
      "           StrelkaHardMemLimit:             \"String\"\n",
      "                 StrelkaScript:               \"File\"\n",
      "           StrelkaSoftMemLimit:             \"String\"\n",
      "                StrelkaThreads:             \"String\"\n",
      "                 StrelkaVcfBgz:               \"File\"\n",
      "              StrelkaVcfBgzTbi:               \"File\"\n",
      "                TrimEnvProfile:               \"File\"\n",
      "              TrimHardMemLimit:             \"String\"\n",
      "                 TrimSeqScript:               \"File\"\n",
      "              TrimSoftMemLimit:             \"String\"\n",
      "                      Trimming:            \"Boolean\"\n",
      "                     TumorBais:               \"File\"\n",
      "                     TumorBams:               \"File\"\n",
      "                          Vqsr:            \"Boolean\"\n",
      "                VqsrEnvProfile:               \"File\"\n",
      "              VqsrHardMemLimit:             \"String\"\n",
      "       VqsrIndelResourceString:             \"String\"\n",
      "                    VqsrScript:               \"File\"\n",
      "         VqsrSnpResourceString:             \"String\"\n",
      "              VqsrSoftMemLimit:             \"String\"\n",
      "                  WorkflowJson:               \"File\"\n"
     ]
    }
   ],
   "source": [
    "def get_wdl_variables_dict(wdl_directory=None):\n",
    "    \"\"\" Usage  config_orderd_dict = get_wdl_variables_dict(wdl_directory)\n",
    "    Get the sorted list of variables and their types from the wdl files in a directory tree\n",
    "    \n",
    "    Args:\n",
    "        wdl_directory:       (default to run directory if not valid directory name)\n",
    "        \n",
    "    Returns:\n",
    "        config_ordered_dict: python OrderedDict of  variable_name: \"type_name\"\n",
    "        \n",
    "    \"\"\"\n",
    "    # Get the complete config.txt file:\n",
    "    skip_line_chars = ['#', '<', '>']\n",
    "    task_types_list = sorted(['Array[File]', 'File', 'File?', 'Boolean', 'String'])\n",
    "    config_vars_dict = {}\n",
    "    for this_dir, dirs, files in os.walk(wdl_directory):\n",
    "        for file in files:\n",
    "            if file[-4:] == '.wdl' and file[0] != '.':\n",
    "                full_filename = os.path.join(this_dir, file)\n",
    "                with open(full_filename, 'r') as fh:\n",
    "                    lines = fh.readlines()\n",
    "                for line in lines:\n",
    "                    l = line.strip()\n",
    "                    if len(l) > 0 and not l[0] in skip_line_chars:\n",
    "                        line_words_list = l.split()\n",
    "                        first_word = line_words_list[0]\n",
    "                        if first_word in task_types_list:\n",
    "                            second_word = line_words_list[1]\n",
    "                            if not second_word in config_vars_dict.keys():\n",
    "                                config_vars_dict[second_word] = '\"' + first_word + '\"'\n",
    "                                \n",
    "                            elif first_word != config_vars_dict[second_word][1:-1]:\n",
    "                                bugger = config_vars_dict[second_word][:-1]\n",
    "                                config_vars_dict[second_word] = bugger + ', ' + first_word  + '\"'\n",
    "                                \n",
    "    config_od = OrderedDict()\n",
    "    for k, v in sorted(config_vars_dict.items()):\n",
    "        config_od[k] = v\n",
    "\n",
    "    return config_od\n",
    "\n",
    "config_vars_dict_ret = get_wdl_variables_dict(wdl_directory)\n",
    "print('found %i variables'%(len(config_vars_dict_ret)))\n",
    "config_od = OrderedDict()\n",
    "for k, v in sorted(config_vars_dict_ret.items()):\n",
    "    config_od[k] = v\n",
    "    \n",
    "for k, v in config_od.items():\n",
    "    print('%30s: %20s'%(k,v))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Walk the 'wdl' directories, find the functional parts, build a comparison spreadsheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_path</th>\n",
       "      <th>imports</th>\n",
       "      <th>calls</th>\n",
       "      <th>workflows</th>\n",
       "      <th>tasks</th>\n",
       "      <th>inputs</th>\n",
       "      <th>outputs</th>\n",
       "      <th>commands</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDL file</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GermlineMasterWorkflow.wdl</th>\n",
       "      <td>../wdl</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SomaticMasterWorkflow.wdl</th>\n",
       "      <td>../wdl</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SomaticVCStandaloneWorkflow.wdl</th>\n",
       "      <td>../wdl</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>manta.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>strelka.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mutect.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>combine_variants.wdl</th>\n",
       "      <td>../wdl/SomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Teststrelka.wdl</th>\n",
       "      <td>../wdl/SomaticVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testcombine_variants.wdl</th>\n",
       "      <td>../wdl/SomaticVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testmutect.wdl</th>\n",
       "      <td>../wdl/SomaticVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deliver_SomaticVC.wdl</th>\n",
       "      <td>../wdl/DeliveryOfSomaticVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdeliver_somatic.wdl</th>\n",
       "      <td>../wdl/DeliveryOfSomaticVC/TestTasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trim_sequences.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dedup.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>merge_aligned_bam.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alignment.wdl</th>\n",
       "      <td>../wdl/Alignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdedup.wdl</th>\n",
       "      <td>../wdl/Alignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Runtrim_sequences.wdl</th>\n",
       "      <td>../wdl/Alignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Runalignment.wdl</th>\n",
       "      <td>../wdl/Alignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TestAlignment.wdl</th>\n",
       "      <td>../wdl/Alignment/Workflow</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deliver_HaplotyperVC.wdl</th>\n",
       "      <td>../wdl/DeliveryOfHaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdeliver_haplotyperVC.wdl</th>\n",
       "      <td>../wdl/DeliveryOfHaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DeliverHaplotyperVC_ForWorkflow.wdl</th>\n",
       "      <td>../wdl/DeliveryOfHaplotyperVC/Workflow</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deliver_alignment.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testdeliver_alignment.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DeliverAlignment_ForWorkflow.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/Workflow</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TestDeliverAlignment.wdl</th>\n",
       "      <td>../wdl/DeliveryOfAlignment/Workflow</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>realignment.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>haplotyper.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Tasks</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testrealignment.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testhaplotyper.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testbqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testvqsr.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/TestTasks</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TestHaplotyperVC.wdl</th>\n",
       "      <td>../wdl/HaplotyperVC/Workflow</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                    src_path  \\\n",
       "WDL file                                                                       \n",
       "GermlineMasterWorkflow.wdl                                            ../wdl   \n",
       "SomaticMasterWorkflow.wdl                                             ../wdl   \n",
       "SomaticVCStandaloneWorkflow.wdl                                       ../wdl   \n",
       "manta.wdl                                             ../wdl/SomaticVC/Tasks   \n",
       "strelka.wdl                                           ../wdl/SomaticVC/Tasks   \n",
       "mutect.wdl                                            ../wdl/SomaticVC/Tasks   \n",
       "combine_variants.wdl                                  ../wdl/SomaticVC/Tasks   \n",
       "Teststrelka.wdl                                   ../wdl/SomaticVC/TestTasks   \n",
       "Testcombine_variants.wdl                          ../wdl/SomaticVC/TestTasks   \n",
       "Testmutect.wdl                                    ../wdl/SomaticVC/TestTasks   \n",
       "deliver_SomaticVC.wdl                       ../wdl/DeliveryOfSomaticVC/Tasks   \n",
       "Testdeliver_somatic.wdl                 ../wdl/DeliveryOfSomaticVC/TestTasks   \n",
       "trim_sequences.wdl                                    ../wdl/Alignment/Tasks   \n",
       "dedup.wdl                                             ../wdl/Alignment/Tasks   \n",
       "merge_aligned_bam.wdl                                 ../wdl/Alignment/Tasks   \n",
       "alignment.wdl                                         ../wdl/Alignment/Tasks   \n",
       "Testdedup.wdl                                     ../wdl/Alignment/TestTasks   \n",
       "Runtrim_sequences.wdl                             ../wdl/Alignment/TestTasks   \n",
       "Runalignment.wdl                                  ../wdl/Alignment/TestTasks   \n",
       "TestAlignment.wdl                                  ../wdl/Alignment/Workflow   \n",
       "deliver_HaplotyperVC.wdl                 ../wdl/DeliveryOfHaplotyperVC/Tasks   \n",
       "Testdeliver_haplotyperVC.wdl         ../wdl/DeliveryOfHaplotyperVC/TestTasks   \n",
       "DeliverHaplotyperVC_ForWorkflow.wdl   ../wdl/DeliveryOfHaplotyperVC/Workflow   \n",
       "deliver_alignment.wdl                       ../wdl/DeliveryOfAlignment/Tasks   \n",
       "Testdeliver_alignment.wdl               ../wdl/DeliveryOfAlignment/TestTasks   \n",
       "DeliverAlignment_ForWorkflow.wdl         ../wdl/DeliveryOfAlignment/Workflow   \n",
       "TestDeliverAlignment.wdl                 ../wdl/DeliveryOfAlignment/Workflow   \n",
       "realignment.wdl                                    ../wdl/HaplotyperVC/Tasks   \n",
       "haplotyper.wdl                                     ../wdl/HaplotyperVC/Tasks   \n",
       "bqsr.wdl                                           ../wdl/HaplotyperVC/Tasks   \n",
       "vqsr.wdl                                           ../wdl/HaplotyperVC/Tasks   \n",
       "Testrealignment.wdl                            ../wdl/HaplotyperVC/TestTasks   \n",
       "Testhaplotyper.wdl                             ../wdl/HaplotyperVC/TestTasks   \n",
       "Testbqsr.wdl                                   ../wdl/HaplotyperVC/TestTasks   \n",
       "Testvqsr.wdl                                   ../wdl/HaplotyperVC/TestTasks   \n",
       "TestHaplotyperVC.wdl                            ../wdl/HaplotyperVC/Workflow   \n",
       "\n",
       "                                     imports  calls  workflows  tasks  inputs  \\\n",
       "WDL file                                                                        \n",
       "GermlineMasterWorkflow.wdl                10     10          1      0      10   \n",
       "SomaticMasterWorkflow.wdl                 11     18          4      0      17   \n",
       "SomaticVCStandaloneWorkflow.wdl            5      7          1      0       6   \n",
       "manta.wdl                                  0      1          0      1       0   \n",
       "strelka.wdl                                0      1          0      1       0   \n",
       "mutect.wdl                                 0      1          0      1       0   \n",
       "combine_variants.wdl                       0      2          0      2       0   \n",
       "Teststrelka.wdl                            1      2          1      0       0   \n",
       "Testcombine_variants.wdl                   1      2          1      0       0   \n",
       "Testmutect.wdl                             1      2          1      0       0   \n",
       "deliver_SomaticVC.wdl                      0      0          1      1       0   \n",
       "Testdeliver_somatic.wdl                    0      0          0      0       0   \n",
       "trim_sequences.wdl                         0      1          0      1       0   \n",
       "dedup.wdl                                  0      1          0      2       1   \n",
       "merge_aligned_bam.wdl                      0      1          0      1       1   \n",
       "alignment.wdl                              0      1          0      1       0   \n",
       "Testdedup.wdl                              1      2          1      0       0   \n",
       "Runtrim_sequences.wdl                      1      3          1      0       3   \n",
       "Runalignment.wdl                           1      2          1      1       3   \n",
       "TestAlignment.wdl                          4      5          1      0       3   \n",
       "deliver_HaplotyperVC.wdl                   0      0          1      1       0   \n",
       "Testdeliver_haplotyperVC.wdl               1      2          1      0       0   \n",
       "DeliverHaplotyperVC_ForWorkflow.wdl        1      1          1      0       1   \n",
       "deliver_alignment.wdl                      0      0          1      1       0   \n",
       "Testdeliver_alignment.wdl                  1      2          1      0       0   \n",
       "DeliverAlignment_ForWorkflow.wdl           1      1          1      0       1   \n",
       "TestDeliverAlignment.wdl                   1      1          1      0       0   \n",
       "realignment.wdl                            0      1          0      1       0   \n",
       "haplotyper.wdl                             0      1          0      2       0   \n",
       "bqsr.wdl                                   0      1          0      2       1   \n",
       "vqsr.wdl                                   0      1          0      2       1   \n",
       "Testrealignment.wdl                        1      1          1      0       0   \n",
       "Testhaplotyper.wdl                         1      2          1      0       0   \n",
       "Testbqsr.wdl                               1      2          1      0       0   \n",
       "Testvqsr.wdl                               1      2          1      0       0   \n",
       "TestHaplotyperVC.wdl                       4      4          1      0       3   \n",
       "\n",
       "                                     outputs  commands  \n",
       "WDL file                                                \n",
       "GermlineMasterWorkflow.wdl                 0         0  \n",
       "SomaticMasterWorkflow.wdl                  0         0  \n",
       "SomaticVCStandaloneWorkflow.wdl            0         0  \n",
       "manta.wdl                                  2         1  \n",
       "strelka.wdl                                1         1  \n",
       "mutect.wdl                                 1         1  \n",
       "combine_variants.wdl                       1         1  \n",
       "Teststrelka.wdl                            1         0  \n",
       "Testcombine_variants.wdl                   1         0  \n",
       "Testmutect.wdl                             1         0  \n",
       "deliver_SomaticVC.wdl                      0         1  \n",
       "Testdeliver_somatic.wdl                    0         0  \n",
       "trim_sequences.wdl                         1         1  \n",
       "dedup.wdl                                  1         1  \n",
       "merge_aligned_bam.wdl                      1         1  \n",
       "alignment.wdl                              1         1  \n",
       "Testdedup.wdl                              0         0  \n",
       "Runtrim_sequences.wdl                      2         0  \n",
       "Runalignment.wdl                           2         0  \n",
       "TestAlignment.wdl                          1         0  \n",
       "deliver_HaplotyperVC.wdl                   0         1  \n",
       "Testdeliver_haplotyperVC.wdl               0         0  \n",
       "DeliverHaplotyperVC_ForWorkflow.wdl        0         0  \n",
       "deliver_alignment.wdl                      1         1  \n",
       "Testdeliver_alignment.wdl                  0         0  \n",
       "DeliverAlignment_ForWorkflow.wdl           0         0  \n",
       "TestDeliverAlignment.wdl                   0         0  \n",
       "realignment.wdl                            1         1  \n",
       "haplotyper.wdl                             1         1  \n",
       "bqsr.wdl                                   1         1  \n",
       "vqsr.wdl                                   1         1  \n",
       "Testrealignment.wdl                        0         0  \n",
       "Testhaplotyper.wdl                         0         0  \n",
       "Testbqsr.wdl                               0         0  \n",
       "Testvqsr.wdl                               0         0  \n",
       "TestHaplotyperVC.wdl                       0         0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_wdl_files_spreadsheet(wdl_directory=None):\n",
    "    \"\"\" wdl_files_spreadsheet = get_wdl_files_spreadsheet(wdl_directory)\n",
    "    get the section statistics spreadsheet for a directory tree of wdl files \n",
    "    \n",
    "    Args:\n",
    "        wdl_directory:       (default to run directory if not valid directory name)\n",
    "        \n",
    "    Returns:\n",
    "        wdl_files_spreadsheet: pandas dataframe with rows of wdl filenames, cols of wdl keywords\n",
    "\n",
    "    \"\"\"\n",
    "    if not wdl_directory is None and os.path.isdir(wdl_directory):\n",
    "        may_dir = wdl_directory\n",
    "    else:\n",
    "        may_dir = os.getcwd()\n",
    "    \n",
    "    dir_name_stripper, _ = os.path.split(wdl_directory)\n",
    "\n",
    "    prefix_ignores = ['.', '_']\n",
    "    \n",
    "    # assemble dictionary - wdl file names: full path  -------- Build empty pandas dataframe\n",
    "    mdl_dict = {}\n",
    "    for dir_name, dir_list, files_list in os.walk(may_dir):\n",
    "        if len(files_list) > 0:\n",
    "            for file_name in files_list:\n",
    "                if file_name[-4:] == '.wdl' and file_name[0] not in prefix_ignores:\n",
    "                    mdl_dict[file_name] = os.path.join(dir_name, file_name)\n",
    "    wdl_df = pd.DataFrame.from_dict(mdl_dict, orient='index', columns=['src_path'])\n",
    "    wdl_df.index.name = 'WDL file'\n",
    "    \n",
    "    # define the rest of the columns in terms of wdl keywords\n",
    "    srch_dict = OrderedDict([('imports','import'), \n",
    "                             ('calls','call'), \n",
    "                             ('workflows','workflow'), \n",
    "                             ('tasks','task'), \n",
    "                             ('inputs','input'), \n",
    "                             ('outputs','output'), \n",
    "                             ('commands', 'command')])\n",
    "    for col_name, _ in srch_dict.items():\n",
    "        wdl_df[col_name] = 0\n",
    "\n",
    "    # for each wdl file: count keywords, get top path ---------- fill pandas dataframe\n",
    "    for name, row in wdl_df.iterrows():\n",
    "        \"\"\" using row.src_path to open files and name to refer to row in dataframe \"\"\"\n",
    "        lines = []\n",
    "        try:\n",
    "            with open(row.src_path, 'r') as fh:\n",
    "                lines = fh.readlines()\n",
    "        except:\n",
    "            print('Fails to open:\\n', row.src_path)\n",
    "            pass\n",
    "        \n",
    "        # count & insert occurrence of keywords\n",
    "        if len(lines) > 0:\n",
    "            for l in lines:\n",
    "                for col_name, key_word in srch_dict.items():\n",
    "                    if key_word in l:\n",
    "                        wdl_df.loc[name,col_name] += 1\n",
    "\n",
    "        # replace the full path name with top path name for readability\n",
    "        full_path, _ = os.path.split(mdl_dict[name])\n",
    "        wdl_df['src_path'].loc[name] = full_path.replace(dir_name_stripper, '..')\n",
    "        \n",
    "    return wdl_df\n",
    "\n",
    "wdl_df = get_wdl_files_spreadsheet(wdl_directory)\n",
    "print(wdl_df.shape)\n",
    "wdl_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read a \".json\" file, build a spreadsheet from a dictionary of variables (2-do add the types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spreadsheet_size =  (0, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variable Value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variable Name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Variable Value]\n",
       "Index: []"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import yaml\n",
    "\n",
    "def read_json_to_df(full_filename):\n",
    "    \"\"\" Usage: json_df = read_json_to_df(full_filename)\n",
    "    read a json file into a spreadsheet\n",
    "    \"\"\"\n",
    "    if os.path.isfile(full_filename) != True:\n",
    "        run_parameters = {}\n",
    "    else:\n",
    "        with open(full_filename, 'r') as file_handle:\n",
    "            run_parameters = yaml.load(file_handle)\n",
    "        if not isinstance(run_parameters, dict):\n",
    "            run_parameters = {}\n",
    "            \n",
    "    json_df = pd.DataFrame.from_dict(run_parameters, orient='index', columns=['Variable Value'])\n",
    "    json_df.index.name = 'Variable Name'\n",
    "    \n",
    "    return json_df\n",
    "\n",
    "config_base = clone_base\n",
    "try_dir = os.path.join(config_base, 'Jsons/GermlineMasterWorkflow.FilledIn.json')\n",
    "jojo_df = read_json_to_df(try_dir)\n",
    "print('spreadsheet_size = ', jojo_df.shape)\n",
    "jojo_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example cell how to develop a python test script\n",
    "    * (cell below is not actually code)\n",
    "    * the cell below would be writte to the desired spot\n",
    "    * could be run from the command line or in the notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "%%writefile check_test_ouput.py\n",
    "\"\"\"\n",
    "check_test_ouput.py\n",
    "Usage:\n",
    "python ~/python/check_test_ouput.py -d /projects/mgc/Project_1/DEL/MVP/cromwell-executions/GermlineMasterWF/\n",
    "or \n",
    "python ~/python/check_test_ouput.py -d `pwd`\n",
    "\n",
    "check return codes and std_out in directory tree\n",
    "\"\"\"\n",
    "\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "good_return_codes_list = ['0', '0\\n']\n",
    "file_names_to_show_list = ['rc', 'stdout', 'stderr']\n",
    "\n",
    "def check_rc_codes(x_directory=None, show_stds=False):\n",
    "    if not x_directory is None and os.path.isdir(x_directory):\n",
    "        dir_tree_root = x_directory\n",
    "    else:\n",
    "        dir_tree_root = os.getcwd()\n",
    "    root_trim_str, _ = os.path.split(dir_tree_root)\n",
    "    for dir_name, dir_list, files_list in os.walk(dir_tree_root):\n",
    "        for filename in files_list:\n",
    "            if filename in file_names_to_show_list:\n",
    "                full_filename = os.path.join(dir_name, filename)\n",
    "                with open(full_filename, 'r') as fh:\n",
    "                    lines = fh.readlines()\n",
    "                if len(lines) <= 0:\n",
    "                    continue\n",
    "                elif filename == 'rc':\n",
    "                    if lines[0] in good_return_codes_list:\n",
    "                        top_dir = dir_name.replace(root_trim_str, '..')\n",
    "                        print('good rc:  %s'%(top_dir))\n",
    "                    else:\n",
    "                        print('\\n\\tBad Dog! Bad Dog!')\n",
    "                        print('code = %s in \\n%s\\n'%(str(lines[0]).strip(), full_filename))\n",
    "                elif show_stds:\n",
    "                    # opportunity to reformat for readability\n",
    "                    for line in lines:\n",
    "                        print(line)\n",
    "                    \n",
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('-d', type=str)\n",
    "    parser.add_argument('-show_stds', nargs=\"?\", const=False, type=bool)\n",
    "    args = parser.parse_args()\n",
    "    check_rc_codes(args.d, args.show_stds)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
