{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multisample branch: Master Workflow for single-normal multi-tumor\n",
    "\n",
    "## Availalbe multilane workflows ( Array[Array[File]] )  in ../src/wdl/\n",
    "* trimming (Optional)\n",
    "    * Inputs are multi-lane fasta\n",
    "    * Alignment/Workflow/Runtrim_sequences.wdl\n",
    "    * Outputs are multi-lane\n",
    "* alignment\n",
    "    * Input option (select_first) inputs are multi-lane fasta\n",
    "    * Alignment/Workflow/Runalignment.wdl\n",
    "    * Outputs are multi-lane fasta\n",
    "* merge\n",
    "    * inputs are multi-lane fasta\n",
    "    * Alignment/Workflow/RunMergeAligned.wdl\n",
    "    * Outputs: Bam, Bai\n",
    "\n",
    "## Available (Bam-Bai) merged functions\n",
    "* dedup (Optional)\n",
    "    * Inputs are Bam, Bai\n",
    "    * Alignment/Tasks/dedup.wdl\n",
    "    * Outputs: - Bam, Bai\n",
    "* deliver alignment\n",
    "    * Input option (select_first) Bam, Bai\n",
    "    * DeliveryOfAlignment/Tasks/deliver_alignment.wdl\n",
    "    * Output - script delivers to:              DeliveryFolder_Alignment\n",
    "* realignment\n",
    "    * Inputs: DeliverAlignOutputBams, DeliverAlignOutputBais\n",
    "    * HaplotyperVC/Tasks/realignment.wdl\n",
    "    * Outputs: OutputBams, OutPutBais\n",
    "* Bqsr (Optional)\n",
    "    * Inputs OutputBams, OutPutBais\n",
    "    * HaplotyperVC/Tasks/bqsr.wdl\n",
    "    * Outputs RecalTable, RecalibratedBam, RecalibratedBai\n",
    "* haplotyper\n",
    "    * Inputs (select_first) and RecalTable\n",
    "    * HaplotyperVC/Tasks/haplotyper.wdl\n",
    "    * Outputs: OutputVcf, OutputVcfIdx\n",
    "* Vqsr (Optional)\n",
    "    * Inputs: OutputVcf, OutputVcfIdx\n",
    "    * HaplotyperVC/Tasks/vqsr.wdl\n",
    "    * Outputs: OutputVcf, OutputVcfIdx\n",
    "* deliver HaplotyperVC\n",
    "    * Inputs: (select_first)\n",
    "    * DeliveryOfHaplotyperVC/Tasks/deliver_HaplotyperVC.wdl\n",
    "    * Outputs: script delivery to DeliveryFolder_HaplotyperVC\n",
    "****"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multi-sample, single normal workflow strategy\n",
    "## Reuse the workflow of multilane workflows through merge\n",
    "* call once with normal\n",
    "* scatter with multisamples\n",
    "## Use merged workflow *SomaticVCStandaloneWorkflow.wdl*\n",
    "* scatter each tumor with normal\n",
    "* delivery task: what is the form of final output ?\n",
    "****"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "###############################################################################################\n",
    "##\t\tSomaticMasterMultiWorkflow.wdl        ##\n",
    "###############################################################################################\n",
    "\n",
    "\n",
    "import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunTrimSequences_MultiSample.wdl\" as TRIM_MULTI_SAMPLE_AND_LANE\n",
    "import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunAlignment_MultiSample.wdl\" as ALIGN_MULTI_SAMPLE_AND_LANE\n",
    "import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunMergeAligned_MultiSample.wdl\" as MERGE_MULTI_SAMPLE\n",
    "\n",
    "workflow SomaticVCMultiSampleMasterWF {\n",
    "        \n",
    "   Array[Array[Array[File]]] InputReads \n",
    "\t\n",
    "   call TRIM_MULTI_SAMPLE_AND_LANE.RunMultiSample_Trimming as TrimMultiSampleMultiLane {\n",
    "      input:\n",
    "         InputReads = InputReads\t\n",
    "   }\n",
    "\n",
    "   call ALIGN_MULTI_SAMPLE_AND_LANE.RunMultiSample_Alignment as AlignMultiSampleMultiLane {\n",
    "      input:\n",
    "         InputReads = TrimMultiSampleMultiLane.Outputs\n",
    "   }\n",
    "\n",
    "   call MERGE_MULTI_SAMPLE.RunMultiSampleMerge as MergeMultiSample {\n",
    "      input:\n",
    "         AlignedMultiSampleOutputBams = AlignMultiSampleMultiLane.OutputBams,\n",
    "         AlignedMultiSampleOutputBais = AlignMultiSampleMultiLane.OutputBais\n",
    "   }\n",
    "}\n",
    "```\n",
    "****\n",
    "## dedup & deliver -- then scatter pairs on *SomaticVCStandaloneWorkflow.wdl*\n",
    "****"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "###############################################################################################\n",
    "##\t\tSomaticMultiSample.wdl       ##\n",
    "###############################################################################################\n",
    "\n",
    "\n",
    "# import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunTrimSequences_MultiSample.wdl\" as TRIM_MULTI_SAMPLE_AND_LANE\n",
    "import \"MayomicsVC/src/wdl/Alignment/Workflow/Runtrim_sequences.wdl\" as TRIM_MULTILANE\n",
    "#import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunAlignment_MultiSample.wdl\" as ALIGN_MULTI_SAMPLE_AND_LANE\n",
    "import \"MayomicsVC/src/wdl/Alignment/Workflow/Runalignment.wdl\" as ALIGN_MULTILANE\n",
    "# import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunMergeAligned_MultiSample.wdl\" as MERGE_MULTI_SAMPLE\n",
    "import \"MayomicsVC/src/wdl/Alignment/Workflow/RunMergeAligned.wdl\" as MERGE_MULTI\n",
    "\n",
    "workflow SomaticVCWorkflow {\n",
    "\n",
    "   Array[Array[File]] NormalInputReads\n",
    "   Array[Array[Array[File]]] TumorInputReads\n",
    "   Array[String] PairSamplesName\n",
    "\n",
    "   full_idx = length(TumorInputReads)\n",
    "   Array[Int] Index = range(full_idx)\n",
    "\n",
    "   scatter (idx in Index) {\n",
    "   \n",
    "      call TRIM_MULTILANE.RunTrimSequencesTask as TRIM_MULTISAMPLE {\n",
    "         input:\n",
    "            PairedEnd = PairedEnd,\n",
    "            InputReads = InputReads[idx],\n",
    "            SampleName = SampleName[idx]\n",
    "      }\n",
    "      \n",
    "      # select_first(flatten(...\n",
    "      call ALIGN_MULTILANE.RunAlignmentTask as ALIGN_MULTISAMPLE {\n",
    "         input:\n",
    "            PairedEnd = PairedEnd,\n",
    "            InputReads = InputReads[idx],\n",
    "            SampleName = SampleName[idx],\n",
    "            PlatformUnit = PlatformUnit\n",
    "      }\n",
    "      # flatten\n",
    "      call MERGE_MULTI.RunMergeAlignedBamsTask as MERGE_MULTI {\n",
    "         input:\n",
    "            SampleName = SampleName[idx],\n",
    "            AlignedOutputBams = AlignedMultiSampleOutputBams[idx],\n",
    "            AlignedOutputBais = AlignedMultiSampleOutputBais[idx]\n",
    "      }\n",
    "      # flatten\n",
    "      # if dedup - dedup\n",
    "      # select_first(flatten(...\n",
    "      # deliver alaignment\n",
    "      #      \n",
    "      #      begin with bam-bai stuff -------------------------------------\n",
    "   }\n",
    "\n",
    "#    call TRIM_MULTI_SAMPLE_AND_LANE.RunMultiSample_Trimming as TrimMultiSampleMultiLane {\n",
    "#       input:\n",
    "#          InputReads = InputReads\t\n",
    "#    }\n",
    "\n",
    "#    call ALIGN_MULTI_SAMPLE_AND_LANE.RunMultiSample_Alignment as AlignMultiSampleMultiLane {\n",
    "#       input:\n",
    "#          InputReads = TrimMultiSampleMultiLane.Outputs\n",
    "#    }\n",
    "\n",
    "#    call MERGE_MULTI_SAMPLE.RunMultiSampleMerge as MergeMultiSample {\n",
    "#       input:\n",
    "#          AlignedMultiSampleOutputBams = AlignMultiSampleMultiLane.OutputBams,\n",
    "#          AlignedMultiSampleOutputBais = AlignMultiSampleMultiLane.OutputBais\n",
    "#    }\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
