{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multisample branch: Master Workflow for single-normal multi-tumor\n",
    "\n",
    "## Availalbe multilane workflows ( Array[Array[File]] )  in ../src/wdl/\n",
    "* trimming (Optional)\n",
    "    * Inputs are multi-lane fasta\n",
    "    * Alignment/Workflow/Runtrim_sequences.wdl\n",
    "    * Outputs are multi-lane\n",
    "* alignment\n",
    "    * Input option (select_first) inputs are multi-lane fasta\n",
    "    * Alignment/Workflow/Runalignment.wdl\n",
    "    * Outputs are multi-lane fasta\n",
    "* merge\n",
    "    * inputs are multi-lane fasta\n",
    "    * Alignment/Workflow/RunMergeAligned.wdl\n",
    "    * Outputs: Bam, Bai\n",
    "\n",
    "## Available (Bam-Bai) merged functions\n",
    "* dedup (Optional)\n",
    "    * Inputs are Bam, Bai\n",
    "    * Alignment/Tasks/dedup.wdl\n",
    "    * Outputs: - Bam, Bai\n",
    "* deliver alignment\n",
    "    * Input option (select_first) Bam, Bai\n",
    "    * DeliveryOfAlignment/Tasks/deliver_alignment.wdl\n",
    "    * Output - script delivers to:              DeliveryFolder_Alignment\n",
    "* realignment\n",
    "    * Inputs: DeliverAlignOutputBams, DeliverAlignOutputBais\n",
    "    * HaplotyperVC/Tasks/realignment.wdl\n",
    "    * Outputs: OutputBams, OutPutBais\n",
    "* Bqsr (Optional)\n",
    "    * Inputs OutputBams, OutPutBais\n",
    "    * HaplotyperVC/Tasks/bqsr.wdl\n",
    "    * Outputs RecalTable, RecalibratedBam, RecalibratedBai\n",
    "* haplotyper\n",
    "    * Inputs (select_first) and RecalTable\n",
    "    * HaplotyperVC/Tasks/haplotyper.wdl\n",
    "    * Outputs: OutputVcf, OutputVcfIdx\n",
    "* Vqsr (Optional)\n",
    "    * Inputs: OutputVcf, OutputVcfIdx\n",
    "    * HaplotyperVC/Tasks/vqsr.wdl\n",
    "    * Outputs: OutputVcf, OutputVcfIdx\n",
    "* deliver HaplotyperVC\n",
    "    * Inputs: (select_first)\n",
    "    * DeliveryOfHaplotyperVC/Tasks/deliver_HaplotyperVC.wdl\n",
    "    * Outputs: script delivery to DeliveryFolder_HaplotyperVC\n",
    "****"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Multi-sample, single normal workflow strategy\n",
    "## Reuse the workflow of multilane workflows through merge\n",
    "* call once with normal\n",
    "* scatter with multisamples\n",
    "## Use merged workflow *SomaticVCStandaloneWorkflow.wdl*\n",
    "* scatter each tumor with normal\n",
    "* delivery task: what is the form of final output ?\n",
    "****"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "###############################################################################################\n",
    "##\t\tSomaticMasterMultiWorkflow.wdl        ##\n",
    "###############################################################################################\n",
    "\n",
    "\n",
    "import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunTrimSequences_MultiSample.wdl\" as TRIM_MULTI_SAMPLE_AND_LANE\n",
    "import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunAlignment_MultiSample.wdl\" as ALIGN_MULTI_SAMPLE_AND_LANE\n",
    "import \"MayomicsVC/src/wdl/SomaticVC/Workflow/RunMergeAligned_MultiSample.wdl\" as MERGE_MULTI_SAMPLE\n",
    "\n",
    "workflow SomaticVCMultiSampleMasterWF {\n",
    "        \n",
    "   Array[Array[Array[File]]] InputReads \n",
    "\t\n",
    "   call TRIM_MULTI_SAMPLE_AND_LANE.RunMultiSample_Trimming as TrimMultiSampleMultiLane {\n",
    "      input:\n",
    "         InputReads = InputReads\t\n",
    "   }\n",
    "\n",
    "   call ALIGN_MULTI_SAMPLE_AND_LANE.RunMultiSample_Alignment as AlignMultiSampleMultiLane {\n",
    "      input:\n",
    "         InputReads = TrimMultiSampleMultiLane.Outputs\n",
    "   }\n",
    "\n",
    "   call MERGE_MULTI_SAMPLE.RunMultiSampleMerge as MergeMultiSample {\n",
    "      input:\n",
    "         AlignedMultiSampleOutputBams = AlignMultiSampleMultiLane.OutputBams,\n",
    "         AlignedMultiSampleOutputBais = AlignMultiSampleMultiLane.OutputBais\n",
    "   }\n",
    "}\n",
    "```\n",
    "****\n",
    "## dedup & deliver -- then scatter pairs on *SomaticVCStandaloneWorkflow.wdl*\n",
    "****"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "###############################################################################################\n",
    "##              SomaticMultiSample.wdl                                                               ##\n",
    "###############################################################################################\n",
    "import \"MayomicsVC/src/wdl/Alignment/Workflow/Runtrim_sequences.wdl\" as CUTADAPTTRIM\n",
    "import \"MayomicsVC/src/wdl/Alignment/Workflow/Runalignment.wdl\" as ALIGNMENT\n",
    "import \"MayomicsVC/src/wdl/Alignment/Tasks/merge_aligned_bam.wdl\" as MERGEBAM\n",
    "import \"MayomicsVC/src/wdl/Alignment/Tasks/dedup.wdl\" as DEDUP\n",
    "import \"MayomicsVC/src/wdl/DeliveryOfAlignment/Tasks/deliver_alignment.wdl\" as DELIVER_Alignment\n",
    "\n",
    "workflow SomaticVCWorkflow {\n",
    "\n",
    "   Array[Array[File]] NormalInputReads\n",
    "   String SampleNameNormal\n",
    "   Array[Array[Array[File]]] TumorInputReads\n",
    "   Array[String] PairSamplesName\n",
    "   Array[Int] Index = range(length(TumorInputReads))\n",
    "\n",
    "   Boolean PairedEnd\n",
    "   Boolean Trimming\n",
    "   Boolean Manta\n",
    "   Boolean MarkDuplicates\n",
    "   Boolean Bqsr\n",
    "   Boolean Vqsr\n",
    "\n",
    "   ######                               Normal sample      ######\n",
    "   ##############################################################\n",
    "   if(Trimming) {\n",
    "      call CUTADAPTTRIM.RunTrimSequencesTask as NormalTrimseq {\n",
    "         input:\n",
    "            InputReads = NormalInputReads\n",
    "      }\n",
    "   }\n",
    "\n",
    "   Array[Array[File]] NormalAlignInputReads = select_first( [NormalTrimseq.Outputs, NormalInputReads] )\n",
    "\n",
    "   call ALIGNMENT.RunAlignmentTask as NormalAlign {\n",
    "      input:\n",
    "         SampleName = SampleNameNormal,\n",
    "         InputReads = NormalAlignInputReads\n",
    "   }\n",
    "\n",
    "   call MERGEBAM.mergebamTask as NormalMergeBam {\n",
    "      input:\n",
    "         SampleName = SampleNameNormal,\n",
    "         InputBams = NormalAlign.OutputBams,\n",
    "         InputBais = NormalAlign.OutputBais\n",
    "   }\n",
    "\n",
    "   if(MarkDuplicates) {\n",
    "\n",
    "       call DEDUP.dedupTask as NormalDedup {\n",
    "          input:\n",
    "             SampleName = SampleNameNormal,\n",
    "             InputBams = NormalMergeBam.OutputBams,\n",
    "             InputBais = NormalMergeBam.OutputBais\n",
    "       }\n",
    "   }\n",
    "\n",
    "   File DeliverNormalAlignOutputBams = select_first( [NormalDedup.OutputBams, NormalMergeBam.OutputBams] )\n",
    "   File DeliverNormalAlignOutputBais = select_first( [NormalDedup.OutputBais, NormalMergeBam.OutputBais] )\n",
    "\n",
    "\n",
    "   call DELIVER_Alignment.deliverAlignmentTask as NormalDAB {\n",
    "      input:\n",
    "         SampleName = SampleNameNormal,\n",
    "         InputBams = DeliverNormalAlignOutputBams,\n",
    "         InputBais = DeliverNormalAlignOutputBais\n",
    "   }\n",
    "   \n",
    "   ##############################################################\n",
    "   ######                scatter block: Tumor samples      ######\n",
    "\n",
    "   scatter (idx in Index) {\n",
    "\n",
    "      String SampleName = PairSamplesName[idx]\n",
    "\n",
    "      if(Trimming) {\n",
    "          call CUTADAPTTRIM.RunTrimSequencesTask as TumorTrimseq {\n",
    "             input:\n",
    "                PairedEnd = PairedEnd,\n",
    "                InputReads = TumorInputReads[idx],\n",
    "                SampleName = SampleName\n",
    "          }\n",
    "      }\n",
    "\n",
    "      Array[Array[File]] TumorAlignInputReads = select_first([TumorTrimseq.Outputs, TumorInputReads[idx]])\n",
    "\n",
    "      call ALIGNMENT.RunAlignmentTask as TumorAlign {\n",
    "         input:\n",
    "            PairedEnd = PairedEnd,\n",
    "            InputReads = TumorAlignInputReads,\n",
    "            SampleName = SampleName\n",
    "      }\n",
    "\n",
    "      call MERGEBAM.mergebamTask as TumorMergeBam {\n",
    "         input:\n",
    "            SampleName = SampleName,\n",
    "            InputBams = TumorAlign.OutputBams,\n",
    "            InputBais = TumorAlign.OutputBais\n",
    "      }\n",
    "\n",
    "      if(MarkDuplicates) {\n",
    "\n",
    "         call DEDUP.dedupTask as TumorDedup {\n",
    "            input:\n",
    "               SampleName = SampleName,\n",
    "               InputBams = TumorMergeBam.OutputBams,\n",
    "               InputBais = TumorMergeBam.OutputBais\n",
    "          }\n",
    "      }\n",
    "\n",
    "      File DeliverTumorAlignOutputBams = select_first([TumorDedup.OutputBams, TumorMergeBam.OutputBams])\n",
    "      File DeliverTumorAlignOutputBais = select_first([TumorDedup.OutputBais, TumorMergeBam.OutputBais])\n",
    "\n",
    "      call DELIVER_Alignment.deliverAlignmentTask as TumorDAB {\n",
    "         input:\n",
    "            SampleName = SampleName,\n",
    "            InputBams = DeliverNormalAlignOutputBams,\n",
    "            InputBais = DeliverNormalAlignOutputBais\n",
    "      }\n",
    "   }\n",
    "}\n",
    "   \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## New and Improved Version # 3\n",
    "* how to switch on index + 1 etc...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
