{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"images/HookahCaterpillerAliceInWDLand.jpg\" width=\"200\" height=\"300\"> <br>\n",
    "# COURT OF aggravated PROGRAMING\n",
    "\n",
    "                v\n",
    "                \n",
    "# MAYOMICSVC SRC PYTHON CONFIG PARSER\n",
    "\n",
    "                aka \n",
    "                \n",
    "# CONFIG PARSER PARSING et al.\n",
    "\n",
    "        I, the undersigned complainant being duly sworn state the following is true and correct \n",
    "        to the best of my knowledge and belief.\n",
    "\n",
    "#### Count One        \n",
    "    From on or about November, 2018 to the present in the district of UIUC, in or about the offices NCSA, the above named code did proceed to aggravate, irritate, insult, inflame and annoy hapless techno-peasents endevoring to translate friviolous file formats in the employ of responsible parties.\n",
    "\n",
    "#### Count Two\n",
    "    Violation of Object Oriented Principles - \"Object\" in question is best not an extensible data type.\n",
    "    * Modular - NOT. Scattered mess of files is not modularity\n",
    "    * Reusability - NOT. Not even useable\n",
    "    * Maintainability - REALLY NOT. Spaghetti code speaks for itself.\n",
    "    * Polymorphism - NOT. Special case program - not adaptable.\n",
    "\n",
    "#### Count Three\n",
    "    Violation of statutes by include validator, key_types, magic-strings and numerous others.\n",
    "    * key_types.json is redundant - contrary to WOMTOOL generated template.json concept.\n",
    "    * contradicts inter-operability standards and standard library functions\n",
    "    \n",
    "##### exempli gratia\n",
    "```json\n",
    "{\n",
    "    \"GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString\": \"\\\"'-M'\\\"\",\n",
    "    \"GermlineMasterWF.haplotype.HaplotyperExtraOptionsString\": \"\\\"'--emit_mode variant --gq_bands 1-60,60-99/19,99 --min_base_qual 10 --pcr_indel_model CONSERVATIVE --phasing 1 --ploidy 2 --prune_factor 2'\\\"\",\n",
    "    \"GermlineMasterWF.vqsr.AnnotateText\": \"\\\"'--annotation DP --annotation QD --annotation FS --annotation SOR --annotation MQ --annotation MQRankSum --annotation ReadPosRankSum'\\\"\",\n",
    "    \"GermlineMasterWF.vqsr.VqsrIndelResourceString\": \"\\\"'--resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf --resource_param dbsnp,known=true,training=false,truth=false,prior=2.0 --resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf --resource_param Mills,known=false,training=true,truth=true,prior=12.0'\\\"\",\n",
    "    \"GermlineMasterWF.vqsr.VqsrSnpResourceString\": \"\\\"'--resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/1000G_phase1.snps.high_confidence.hg38.vcf.gz --resource_param 1000G,known=false,training=true,truth=false,prior=10.0 --resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/1000G_omni2.5.hg38.vcf.gz --resource_param omni,known=false,training=true,truth=false,prior=12.0 --resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf --resource_param dbsnp,known=true,training=false,truth=false,prior=2.0 --resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/hapmap_3.3.hg38.vcf.gz --resource_param hapmap,known=false,training=true,truth=true,prior=15.0'\\\"\"\n",
    "}\n",
    "```\n",
    "##### of templated types:\n",
    "```json\n",
    "{\n",
    "    \"GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString\": \"String\",\n",
    "    \"GermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString\": \"String\",\n",
    "    \"GermlineMasterWF.haplotype.HaplotyperExtraOptionsString\": \"String\",\n",
    "    \"GermlineMasterWF.vqsr.AnnotateText\": \"String\",\n",
    "    \"GermlineMasterWF.vqsr.VqsrIndelResourceString\": \"String\",\n",
    "    \"GermlineMasterWF.vqsr.VqsrSnpResourceString\": \"String\"\n",
    "}\n",
    "```\n",
    "##### primarily foune in: run_info.txt\n",
    "```bash\n",
    "BWAExtraOptionsString=\"\"'-M'\"\"\n",
    "\n",
    "HaplotyperExtraOptionsString=\"\"'--emit_mode variant --gq_bands 1-60,60-99/19,99 --min_base_qual 10 --pcr_indel_model CONSERVATIVE --phasing 1 --ploidy 2 --prune_factor 2'\"\"\n",
    "\n",
    "AnnotateText=\"\"'--annotation DP --annotation QD --annotation FS --annotation SOR --annotation MQ --annotation MQRankSum --annotation ReadPosRankSum'\"\"\n",
    "\n",
    "VqsrIndelResourceString=\"\"'--resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf --resource_param dbsnp,known=true,training=false,truth=false,prior=2.0 --resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf --resource_param Mills,known=false,training=true,truth=true,prior=12.0'\"\"\n",
    "\n",
    "VqsrSnpResourceString=\"\"'--resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/1000G_phase1.snps.high_confidence.hg38.vcf.gz --resource_param 1000G,known=false,training=true,truth=false,prior=10.0 --resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/1000G_omni2.5.hg38.vcf.gz --resource_param omni,known=false,training=true,truth=false,prior=12.0 --resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf --resource_param dbsnp,known=true,training=false,truth=false,prior=2.0 --resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/hapmap_3.3.hg38.vcf.gz --resource_param hapmap,known=false,training=true,truth=true,prior=15.0'\"\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
