{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[json python std lib](https://docs.python.org/3/library/json.html) <br>\n",
    "# Bottom Line:\n",
    "They all open json files as text with or without double quotes. <br>\n",
    "The readlines( ) option drops the newline characters and returns a list. <br>\n",
    "The splitlines( ) option keeps the newline charcters and returns a list. <br>\n",
    "The json.loads( ) removes newline characters and constructs a propper python dict. <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Open the json file - compare methods -- aka first mistake:\n",
    "* output is dictionary of single-quoted strings\n",
    "* the single quotes are superfluous\n",
    "* json.loads as a list of lines or yaml avoids the single quotes.\n",
    "\n",
    "[Current](https://git.ncsa.illinois.edu/mayomics/MayomicsVC/blob/master/src/python/util/util.py) <br>\n",
    "\n",
    "```python\n",
    "import sys\n",
    "import json\n",
    "\n",
    "def read_json_file(json_file, project_logger, json_not_found_error_code, json_bad_format_error_code):\n",
    "    \"\"\"\n",
    "    Reads a JSON formatted file, and returns its contents as a Python dictionary\n",
    "\n",
    "    Inputs:\n",
    "        The json file\n",
    "        The project logger (for error messages to get passed to)\n",
    "        The error code if a JSON file is not found\n",
    "        The error code if a JSON file is improperly formatted\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Open the file\n",
    "        with open(json_file) as F:\n",
    "            # Read the file's contents as a string\n",
    "            json_str = F.read()\n",
    "            # Return the data as a Python dictionary\n",
    "            return json.loads(json_str)\n",
    "    except FileNotFoundError:\n",
    "        project_logger.log_error(\n",
    "            json_not_found_error_code,\n",
    "            'Could not open json file \"' + str(json_file) + '\": JSON file could not be found'\n",
    "        )\n",
    "        sys.exit(1)\n",
    "    except json.decoder.JSONDecodeError:\n",
    "        project_logger.log_error(\n",
    "            json_bad_format_error_code,\n",
    "            'Could not open json file \"' + str(json_file) + '\": JSON file not formatted properly'\n",
    "        )\n",
    "        sys.exit(1)\n",
    "                \n",
    "\n",
    "def read_input_file(self, file_path):\n",
    "    \n",
    "    try:\n",
    "        with open(file_path, \"r\") as F:\n",
    "            return F.read().splitlines()\n",
    "        \n",
    "    except FileNotFoundError:\n",
    "        self.project_logger.log_error(\"E.par.Fil.1\", 'Input file \"' + str(file_path) + \n",
    "                                      '\" could not be found')\n",
    "        sys.exit(1)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "path_to_development_code_modules = '.'\n",
    "sys.path.insert(1, path_to_development_code_modules)\n",
    "from womtool_template_fill_in import *\n",
    "\n",
    "import json\n",
    "\n",
    "def read_raw_hee_haw(file_path):\n",
    "    \n",
    "    with open(file_path, \"r\") as F:\n",
    "        return F.read().splitlines()\n",
    "\n",
    "\n",
    "def read_json_raw(json_full_filename):\n",
    "    \n",
    "    with open(json_full_filename, 'r') as fh:\n",
    "        lines = fh.readlines()\n",
    "        \n",
    "    return lines\n",
    "\n",
    "\n",
    "def dict_from_json_file(json_full_filename):\n",
    "    # lines is list with newline chars left in\n",
    "    lines = read_json_raw(json_full_filename)\n",
    "    S = ''\n",
    "    for line in lines:\n",
    "        S += line.strip()\n",
    "    \n",
    "    return json.loads(S)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "jsonTemplate = '../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json'\n",
    "json_filled_in = '../data/GermlineParameters/Jsons/GermlineMasterWorkflow.BAK.FilledIn.json'\n",
    "\n",
    "#                   Sets all cells on the page:\n",
    "json_compare_file = jsonTemplate\n",
    "print(os.path.isfile(json_compare_file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:\n",
      "../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json\n",
      "with method:\n",
      "read_raw_hee_haw\n",
      "\n",
      "<class 'list'> \n",
      "\n",
      "\n",
      "{\n",
      "  \"GermlineMasterWF.realign.RealignSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.Bqsr\": \"Boolean\",\n",
      "  \"GermlineMasterWF.vqsr.VqsrHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.dedup.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.merge.MergeBamEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.DHVC.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.bqsr.BqsrSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.bqsr.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.CenterName\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.VqsrSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.BqsrHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.AlignEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.vqsr.AnnotateText\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.RefPac\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.bqsr.Ref\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.HaplotyperSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.BqsrKnownSites\": \"String\",\n",
      "  \"GermlineMasterWF.realign.RealignHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.RefFai\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.HaplotyperEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.realign.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.MarkDuplicates\": \"Boolean\",\n",
      "  \"GermlineMasterWF.haplotype.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.RefBwt\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.Platform\": \"String\",\n",
      "  \"GermlineMasterWF.realign.RealignmentScript\": \"File\",\n",
      "  \"GermlineMasterWF.NormalInputReads\": \"Array[Array[File]]\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.realign.RefFai\": \"File\",\n",
      "  \"GermlineMasterWF.merge.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.realign.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.DHVC.WorkflowJson\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.PairedEnd\": \"Boolean\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.merge.MergeBamScript\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.RefSa\": \"File\",\n",
      "  \"GermlineMasterWF.bqsr.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.DedupScript\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.Ref\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.AlignSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.RefFai\": \"File\",\n",
      "  \"GermlineMasterWF.DAB.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.AlignmentScript\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.RefPac\": \"File\",\n",
      "  \"GermlineMasterWF.DAB.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.HaplotyperVCFSourceField\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.DBSNP\": \"File\",\n",
      "  \"GermlineMasterWF.vqsr.VqsrIndelResourceString\": \"String\",\n",
      "  \"GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC\": \"String\",\n",
      "  \"GermlineMasterWF.DAB.DeliveryFolder_Alignment\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.VqsrScript\": \"File\",\n",
      "  \"GermlineMasterWF.align.PlatformUnit\": \"Array[String]\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.RefBwt\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.trimseq.PairedEnd\": \"Boolean\",\n",
      "  \"GermlineMasterWF.haplotype.HaplotyperScript\": \"File\",\n",
      "  \"GermlineMasterWF.Vqsr\": \"Boolean\",\n",
      "  \"GermlineMasterWF.vqsr.VqsrSnpResourceString\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.DedupEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.RefAmb\": \"File\",\n",
      "  \"GermlineMasterWF.realign.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.DAB.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.CenterName\": \"String\",\n",
      "  \"GermlineMasterWF.merge.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.realign.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.Platform\": \"String\",\n",
      "  \"GermlineMasterWF.realign.RealignmentKnownSites\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.BqsrScript\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.AlignHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.DHVC.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.HaplotyperExtraOptionsString\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.VqsrEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.AlignEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSeqScript\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdaptThreads\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.BqsrEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.vqsr.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.Library\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.Ref\": \"File\",\n",
      "  \"GermlineMasterWF.realign.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.PairedEnd\": \"Boolean\",\n",
      "  \"GermlineMasterWF.trimseq.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.realign.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.DAB.DeliveryAlignment_Script\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.HaplotyperHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.realign.Ref\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.Adapters\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.Ref\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.RefSa\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.RefAnn\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads\": \"String\",\n",
      "  \"GermlineMasterWF.merge.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.DedupHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.dedup.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.RefFai\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.RefAnn\": \"File\",\n",
      "  \"GermlineMasterWF.DHVC.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.align.PairedEnd\": \"Boolean\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.Library\": \"String\",\n",
      "  \"GermlineMasterWF.bqsr.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.merge.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.DAB.WorkflowJson\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd\": \"Boolean\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.PairedEnd\": \"Boolean\",\n",
      "  \"GermlineMasterWF.align.SampleName\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.ChunkSizeInBases\": \"String\",\n",
      "  \"GermlineMasterWF.Trimming\": \"Boolean\",\n",
      "  \"GermlineMasterWF.vqsr.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdapt\": \"String\",\n",
      "  \"GermlineMasterWF.haplotype.DBSNPIdx\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.dedup.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.DAB.BashPreamble\": \"File\",\n",
      "  \"GermlineMasterWF.DHVC.BashSharedFunctions\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.DebugMode\": \"String\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.realign.RealignEnvProfile\": \"File\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.RefAmb\": \"File\",\n",
      "  \"GermlineMasterWF.haplotype.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_paired.SentieonThreads\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.AlignmentScript\": \"File\",\n",
      "  \"GermlineMasterWF.merge.MergeHardMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString\": \"String\",\n",
      "  \"GermlineMasterWF.vqsr.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.merge.Sentieon\": \"String\",\n",
      "  \"GermlineMasterWF.merge.MergeSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.align.ALIGN_single.Ref\": \"File\",\n",
      "  \"GermlineMasterWF.trimseq.TRIMSEQ_paired.Adapters\": \"File\",\n",
      "  \"GermlineMasterWF.dedup.DedupSoftMemLimit\": \"String\",\n",
      "  \"GermlineMasterWF.merge.BashSharedFunctions\": \"File\"\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('File:\\n%s\\nwith method:\\nread_raw_hee_haw\\n'%(json_compare_file))\n",
    "lines = read_raw_hee_haw(json_compare_file)\n",
    "print(type(lines),'\\n\\n')\n",
    "\n",
    "for line in lines:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:\n",
      "../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json\n",
      "with method:\n",
      "read_json_raw\n",
      "\n",
      "\n",
      "<class 'list'> \n",
      "\n",
      "\n",
      "{\n",
      "\"GermlineMasterWF.realign.RealignSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.Bqsr\": \"Boolean\",\n",
      "\"GermlineMasterWF.vqsr.VqsrHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.dedup.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.merge.MergeBamEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.DHVC.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.bqsr.BqsrSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.bqsr.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.CenterName\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.VqsrSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.BqsrHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.AlignEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.vqsr.AnnotateText\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.RefPac\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.bqsr.Ref\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.HaplotyperSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.BqsrKnownSites\": \"String\",\n",
      "\"GermlineMasterWF.realign.RealignHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.dedup.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.dedup.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.RefFai\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.HaplotyperEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.realign.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.MarkDuplicates\": \"Boolean\",\n",
      "\"GermlineMasterWF.haplotype.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.RefBwt\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.Platform\": \"String\",\n",
      "\"GermlineMasterWF.realign.RealignmentScript\": \"File\",\n",
      "\"GermlineMasterWF.NormalInputReads\": \"Array[Array[File]]\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.realign.RefFai\": \"File\",\n",
      "\"GermlineMasterWF.merge.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.realign.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.DHVC.WorkflowJson\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.PairedEnd\": \"Boolean\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.merge.MergeBamScript\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.RefSa\": \"File\",\n",
      "\"GermlineMasterWF.bqsr.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.dedup.DedupScript\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.Ref\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.AlignSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.RefFai\": \"File\",\n",
      "\"GermlineMasterWF.DAB.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.dedup.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.AlignmentScript\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.RefPac\": \"File\",\n",
      "\"GermlineMasterWF.DAB.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.HaplotyperVCFSourceField\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.DBSNP\": \"File\",\n",
      "\"GermlineMasterWF.vqsr.VqsrIndelResourceString\": \"String\",\n",
      "\"GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC\": \"String\",\n",
      "\"GermlineMasterWF.DAB.DeliveryFolder_Alignment\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.VqsrScript\": \"File\",\n",
      "\"GermlineMasterWF.align.PlatformUnit\": \"Array[String]\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.RefBwt\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.trimseq.PairedEnd\": \"Boolean\",\n",
      "\"GermlineMasterWF.haplotype.HaplotyperScript\": \"File\",\n",
      "\"GermlineMasterWF.Vqsr\": \"Boolean\",\n",
      "\"GermlineMasterWF.vqsr.VqsrSnpResourceString\": \"String\",\n",
      "\"GermlineMasterWF.dedup.DedupEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.RefAmb\": \"File\",\n",
      "\"GermlineMasterWF.realign.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.DAB.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.CenterName\": \"String\",\n",
      "\"GermlineMasterWF.merge.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.realign.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.Platform\": \"String\",\n",
      "\"GermlineMasterWF.realign.RealignmentKnownSites\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.BqsrScript\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.AlignHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.DHVC.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.HaplotyperExtraOptionsString\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.VqsrEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.AlignEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSeqScript\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdaptThreads\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.BqsrEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.vqsr.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.Library\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.Ref\": \"File\",\n",
      "\"GermlineMasterWF.realign.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.PairedEnd\": \"Boolean\",\n",
      "\"GermlineMasterWF.trimseq.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.realign.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.DAB.DeliveryAlignment_Script\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.HaplotyperHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.realign.Ref\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.Adapters\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.Ref\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.RefSa\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.RefAnn\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads\": \"String\",\n",
      "\"GermlineMasterWF.merge.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.dedup.DedupHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.dedup.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.RefFai\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.RefAnn\": \"File\",\n",
      "\"GermlineMasterWF.DHVC.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.align.PairedEnd\": \"Boolean\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.Library\": \"String\",\n",
      "\"GermlineMasterWF.bqsr.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.merge.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.DAB.WorkflowJson\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd\": \"Boolean\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.PairedEnd\": \"Boolean\",\n",
      "\"GermlineMasterWF.align.SampleName\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.ChunkSizeInBases\": \"String\",\n",
      "\"GermlineMasterWF.Trimming\": \"Boolean\",\n",
      "\"GermlineMasterWF.vqsr.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdapt\": \"String\",\n",
      "\"GermlineMasterWF.haplotype.DBSNPIdx\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.dedup.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.DAB.BashPreamble\": \"File\",\n",
      "\"GermlineMasterWF.DHVC.BashSharedFunctions\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.DebugMode\": \"String\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.realign.RealignEnvProfile\": \"File\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.RefAmb\": \"File\",\n",
      "\"GermlineMasterWF.haplotype.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_paired.SentieonThreads\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.AlignmentScript\": \"File\",\n",
      "\"GermlineMasterWF.merge.MergeHardMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString\": \"String\",\n",
      "\"GermlineMasterWF.vqsr.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.merge.Sentieon\": \"String\",\n",
      "\"GermlineMasterWF.merge.MergeSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.align.ALIGN_single.Ref\": \"File\",\n",
      "\"GermlineMasterWF.trimseq.TRIMSEQ_paired.Adapters\": \"File\",\n",
      "\"GermlineMasterWF.dedup.DedupSoftMemLimit\": \"String\",\n",
      "\"GermlineMasterWF.merge.BashSharedFunctions\": \"File\"\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('File:\\n%s\\nwith method:\\nread_json_raw\\n\\n'%(json_compare_file))\n",
    "lines = read_json_raw(json_compare_file)\n",
    "print(type(lines),'\\n\\n')\n",
    "\n",
    "for line in lines:\n",
    "    print(line.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:\n",
      "../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json\n",
      "with method:\n",
      "dict_from_json_file\n",
      "\n",
      "<class 'dict'> 165 keys\n",
      "\n",
      "GermlineMasterWF.realign.RealignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.Bqsr:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeBamEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.CenterName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.AnnotateText:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefPac:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrKnownSites:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.MarkDuplicates:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.haplotype.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefBwt:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Platform:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.NormalInputReads:\n",
      "\tArray[Array[File]]\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.WorkflowJson:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeBamScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefSa:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefPac:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperVCFSourceField:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DBSNP:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrIndelResourceString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.DeliveryFolder_Alignment:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.PlatformUnit:\n",
      "\tArray[String]\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefBwt:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.Vqsr:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrSnpResourceString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefAmb:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.CenterName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Platform:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignmentKnownSites:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSeqScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdaptThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Library:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.DeliveryAlignment_Script:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.Adapters:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefSa:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefAnn:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefAnn:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Library:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.WorkflowJson:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.align.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.ChunkSizeInBases:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.Trimming:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdapt:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DBSNPIdx:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.RealignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefAmb:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.MergeSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.Adapters:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.DedupSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.BashSharedFunctions:\n",
      "\tFile\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('File:\\n%s\\nwith method:\\ndict_from_json_file\\n'%(json_compare_file))\n",
    "proppa_dict = {}\n",
    "\n",
    "if os.path.isfile(json_compare_file) == True:\n",
    "    proppa_dict = dict_from_json_file(json_compare_file)\n",
    "    \n",
    "print(type(proppa_dict), len(proppa_dict), 'keys\\n')\n",
    "    \n",
    "for k,v in proppa_dict.items():\n",
    "    print('%s:\\n\\t%s\\n'%(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:\n",
      "../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json\n",
      "with method:\n",
      "get_json_file_dict\n",
      "\n",
      "\n",
      "<class 'dict'>\n",
      "165 variables found\n",
      "\n",
      "<class 'str'>\n",
      "<class 'str'>\n",
      "S \n",
      "\n",
      "\n",
      "GermlineMasterWF.realign.RealignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.Bqsr:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeBamEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.CenterName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.AnnotateText:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefPac:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrKnownSites:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.MarkDuplicates:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.haplotype.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefBwt:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Platform:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.NormalInputReads:\n",
      "\tArray[Array[File]]\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.WorkflowJson:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeBamScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefSa:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefPac:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperVCFSourceField:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DBSNP:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrIndelResourceString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.DeliveryFolder_Alignment:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.PlatformUnit:\n",
      "\tArray[String]\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefBwt:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.Vqsr:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrSnpResourceString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefAmb:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.CenterName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Platform:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignmentKnownSites:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSeqScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdaptThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Library:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.DeliveryAlignment_Script:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.Adapters:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefSa:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefAnn:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefAnn:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Library:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.WorkflowJson:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.align.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.ChunkSizeInBases:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.Trimming:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdapt:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DBSNPIdx:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.RealignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefAmb:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.MergeSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.Adapters:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.DedupSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.BashSharedFunctions:\n",
      "\tFile\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# demonstrate get_run_file_dict on json with quotes:\n",
    "TestTask_dir = os.getcwd()\n",
    "TestTask_jason_file = json_compare_file # json_filled_in #jsonTemplate # 'jjalltheway.json'\n",
    "\n",
    "json_fullfilename = os.path.join(TestTask_dir, TestTask_jason_file)\n",
    "\n",
    "print('File:\\n%s\\nwith method:\\nget_json_file_dict\\n\\n'%(json_compare_file))\n",
    "\n",
    "if os.path.isfile(json_fullfilename):\n",
    "    json_template_dict = get_json_file_dict(json_fullfilename)\n",
    "    print(type(json_template_dict))\n",
    "    print('{0} variables found\\n'.format(len(json_template_dict)))\n",
    "    print(type(list(json_template_dict.keys())[0]))\n",
    "    print(type(list(json_template_dict.values())[0]))\n",
    "    print(list(json_template_dict.values())[0][0],'\\n\\n')\n",
    "    if len(json_template_dict) > 0:\n",
    "        for k, v in json_template_dict.items():\n",
    "            print('%s:\\n\\t%s\\n'%(k,v))\n",
    "else:\n",
    "    print(json_fullfilename, '\\nNot Found')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#                     Implement the current (Feb 2019 gitlab repo code)\n",
    "import sys\n",
    "import json\n",
    "import logging\n",
    "\n",
    "def read_json_file(json_file, \n",
    "                   project_logger=logging.getLogger(name='parse_test'), \n",
    "                   json_not_found_error_code=\"E.par.JSN.1\", \n",
    "                   json_bad_format_error_code=\"E.par.JSN.2\"):\n",
    "    \"\"\" Usage: str_else_crash = read_json_file( json_file, \n",
    "                                                project_logger, \n",
    "                                                json_not_found_error_code, \n",
    "                                                json_bad_format_error_code)\n",
    "    Reads a JSON formatted file, and returns its contents as a Python dictionary\n",
    "\n",
    "    Inputs:\n",
    "        The json file\n",
    "        The project logger (for error messages to get passed to)\n",
    "        The error code if a JSON file is not found\n",
    "        The error code if a JSON file is improperly formatted\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Open the file\n",
    "        with open(json_file) as F:\n",
    "            # Read the file's contents as a string\n",
    "            json_str = F.read()\n",
    "            # Return the data as a Python dictionary\n",
    "            return json.loads(json_str)\n",
    "        \n",
    "    except FileNotFoundError:\n",
    "        project_logger.log_error(\n",
    "            json_not_found_error_code,\n",
    "            'Could not open json file \"' + str(json_file) + '\": JSON file could not be found'\n",
    "        )\n",
    "        sys.exit(1)\n",
    "    except json.decoder.JSONDecodeError:\n",
    "        project_logger.log_error(\n",
    "            json_bad_format_error_code,\n",
    "            'Could not open json file \"' + str(json_file) + '\": JSON file not formatted properly'\n",
    "        )\n",
    "        sys.exit(1)\n",
    "        \n",
    "# project_logger = logging.getLogger(name='any_name')\n",
    "# json_not_found_error_code=\"E.par.JSN.1\",\n",
    "# json_bad_format_error_code=\"E.par.JSN.2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:\n",
      "../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json\n",
      "with method:\n",
      "read_json_file\n",
      "\n",
      "<class 'dict'>\n",
      "165 variables found\n",
      "\n",
      "<class 'str'>\n",
      "<class 'str'>\n",
      "S \n",
      "\n",
      "\n",
      "GermlineMasterWF.realign.RealignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.Bqsr:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeBamEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.CenterName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.AnnotateText:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefPac:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrKnownSites:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.MarkDuplicates:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.haplotype.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefBwt:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Platform:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.NormalInputReads:\n",
      "\tArray[Array[File]]\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.WorkflowJson:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeBamScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefSa:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.bqsr.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefPac:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperVCFSourceField:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DBSNP:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrIndelResourceString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.DeliveryFolder_Alignment:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.PlatformUnit:\n",
      "\tArray[String]\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefBwt:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.Vqsr:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrSnpResourceString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefAmb:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.CenterName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Platform:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.RealignmentKnownSites:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DHVC.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.VqsrEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSeqScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdaptThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BqsrEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.vqsr.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Library:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.DeliveryAlignment_Script:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.HaplotyperHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.realign.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.Adapters:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefSa:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.RefAnn:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.DedupHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.dedup.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.RefFai:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefAnn:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Library:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.bqsr.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.DAB.WorkflowJson:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.PairedEnd:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.align.SampleName:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.ChunkSizeInBases:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.Trimming:\n",
      "\tBoolean\n",
      "\n",
      "GermlineMasterWF.vqsr.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdapt:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.haplotype.DBSNPIdx:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DAB.BashPreamble:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.DHVC.BashSharedFunctions:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.DebugMode:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.TrimEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.realign.RealignEnvProfile:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.RefAmb:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.haplotype.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_paired.SentieonThreads:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.AlignmentScript:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.merge.MergeHardMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.vqsr.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.Sentieon:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.MergeSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.align.ALIGN_single.Ref:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.trimseq.TRIMSEQ_paired.Adapters:\n",
      "\tFile\n",
      "\n",
      "GermlineMasterWF.dedup.DedupSoftMemLimit:\n",
      "\tString\n",
      "\n",
      "GermlineMasterWF.merge.BashSharedFunctions:\n",
      "\tFile\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('File:\\n%s\\nwith method:\\nread_json_file\\n'%(json_compare_file))\n",
    "\n",
    "dict_else_crash = read_json_file(json_compare_file)\n",
    "\n",
    "print(type(dict_else_crash))\n",
    "print('{0} variables found\\n'.format(len(dict_else_crash)))\n",
    "print(type(list(dict_else_crash.keys())[0]))\n",
    "print(type(list(dict_else_crash.values())[0]))\n",
    "print(list(dict_else_crash.values())[0][0],'\\n\\n')\n",
    "\n",
    "for k, v in dict_else_crash.items():\n",
    "    print('%s:\\n\\t%s\\n'%(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File:\n",
      "../data/GermlineParameters/Jsons/GermlineMasterWorkflow.template.json\n",
      "with method:\n",
      "get_config_file_dict\n",
      "\n",
      "empty dictionary :) no \"=\" signs\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "print('File:\\n%s\\nwith method:\\nget_config_file_dict\\n'%(json_compare_file))\n",
    "\n",
    "dict_else_crash = get_config_file_dict(json_compare_file)\n",
    "# dict_else_crash = read_json_file(json_compare_file)\n",
    "if len(dict_else_crash) > 0:\n",
    "    print(type(dict_else_crash))\n",
    "    print('{0} variables found\\n'.format(len(dict_else_crash)))\n",
    "    print(type(list(dict_else_crash.keys())[0]))\n",
    "    print(type(list(dict_else_crash.values())[0]))\n",
    "    print(list(dict_else_crash.values())[0][0],'\\n\\n')\n",
    "\n",
    "    for k, v in dict_else_crash.items():\n",
    "        print('%s:\\n\\t%s\\n'%(k,v))\n",
    "else:\n",
    "    print('empty dictionary :) no \"=\" signs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Counter',\n",
       " 'In',\n",
       " 'OrderedDict',\n",
       " 'Out',\n",
       " 'TestTask_dir',\n",
       " 'TestTask_jason_file',\n",
       " '_',\n",
       " '__',\n",
       " '___',\n",
       " '__builtin__',\n",
       " '__builtins__',\n",
       " '__doc__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__spec__',\n",
       " '_dh',\n",
       " '_i',\n",
       " '_i1',\n",
       " '_i10',\n",
       " '_i2',\n",
       " '_i3',\n",
       " '_i4',\n",
       " '_i5',\n",
       " '_i6',\n",
       " '_i7',\n",
       " '_i8',\n",
       " '_i9',\n",
       " '_ih',\n",
       " '_ii',\n",
       " '_iii',\n",
       " '_oh',\n",
       " 'argparse',\n",
       " 'args_dict_to_filledin_json',\n",
       " 'assemble_config_dict',\n",
       " 'configure_json_dict',\n",
       " 'defaultdict',\n",
       " 'dict_else_crash',\n",
       " 'dict_from_json_file',\n",
       " 'exit',\n",
       " 'get_config_file_dict',\n",
       " 'get_ipython',\n",
       " 'get_json_file_dict',\n",
       " 'get_json_keys_config_dict',\n",
       " 'json',\n",
       " 'jsonTemplate',\n",
       " 'json_compare_file',\n",
       " 'json_filled_in',\n",
       " 'json_fullfilename',\n",
       " 'json_template_dict',\n",
       " 'k',\n",
       " 'line',\n",
       " 'lines',\n",
       " 'logging',\n",
       " 'os',\n",
       " 'parse_args',\n",
       " 'path_to_development_code_modules',\n",
       " 'proppa_dict',\n",
       " 'quit',\n",
       " 'read_json_file',\n",
       " 'read_json_raw',\n",
       " 'read_raw_hee_haw',\n",
       " 'sys',\n",
       " 'v',\n",
       " 'write_filled_in_json_dict']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " key :\tGermlineMasterWF.Bqsr\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.DeliveryAlignment_Script\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/deliver_alignment.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.DeliveryFolder_Alignment\n",
      "value:\t/projects/bioinformatics/DEL/Delivery/Alignment\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DAB.WorkflowJson\n",
      "value:\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.DeliveryFolder_HaplotyperVC\n",
      "value:\t/projects/bioinformatics/DEL/Delivery/HaplotyperVC\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.DeliveryHaplotyperVC_Script\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/deliver_haplotyperVC.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.DHVC.WorkflowJson\n",
      "value:\t/projects/bioinformatics/DEL/Jsons/GermlineMasterWorkflow.FilledIn.json\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.MarkDuplicates\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.NormalInputReads\n",
      "value:\t[['/projects/bioinformatics/DEL/Inputs/WGS_chr1_5X_E0.005_L1_read1.fastq.gz', '/projects/bioinformatics/DEL/Inputs/WGS_chr1_5X_E0.005_L1_read2.fastq.gz'], ['/projects/bioinformatics/DEL/Inputs/WGS_chr1_5X_E0.005_L2_read1.fastq.gz', '/projects/bioinformatics/DEL/Inputs/WGS_chr1_5X_E0.005_L2_read2.fastq.gz']]\n",
      "<class 'str'>\t<class 'list'>\n",
      "\n",
      " key :\tGermlineMasterWF.Trimming\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.Vqsr\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.AlignEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/AlignEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.AlignHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.AlignSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.AlignmentScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/alignment.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.BWAExtraOptionsString\n",
      "value:\t\"'-M'\"\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.CenterName\n",
      "value:\tsome_seq_center\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.ChunkSizeInBases\n",
      "value:\t10000000\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.Library\n",
      "value:\tfake_lib\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.PairedEnd\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.Platform\n",
      "value:\tILLUMINA\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.Ref\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.RefAmb\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.amb\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.RefAnn\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.ann\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.RefBwt\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.bwt\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.RefDict\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.dict\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.RefPac\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.pac\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.RefSa\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.sa\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_paired.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.AlignEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/AlignEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.AlignHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.AlignSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.AlignmentScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/alignment.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.BWAExtraOptionsString\n",
      "value:\t\"'-M'\"\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.CenterName\n",
      "value:\tsome_seq_center\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.ChunkSizeInBases\n",
      "value:\t10000000\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.Library\n",
      "value:\tfake_lib\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.PairedEnd\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.Platform\n",
      "value:\tILLUMINA\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.Ref\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.RefAmb\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.amb\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.RefAnn\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.ann\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.RefBwt\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.bwt\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.RefDict\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.dict\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.RefPac\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.pac\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.RefSa\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.sa\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.ALIGN_single.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.PairedEnd\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.align.PlatformUnit\n",
      "value:\t['FLOWCELL_BARCODE.LANE.SAMPLE_BARCODE', 'FLOWCELL_BARCODE.LANE.SAMPLE_BARCODE']\n",
      "<class 'str'>\t<class 'list'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BqsrEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/BqsrEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BqsrHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BqsrKnownSites\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf,/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BqsrScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/bqsr.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.BqsrSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.Ref\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefAmb\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.amb\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefAnn\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.ann\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefBwt\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.bwt\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefDict\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.dict\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefFai\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.fai\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefPac\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.pac\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.RefSa\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.sa\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.bqsr.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.DedupEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/DedupEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.DedupHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.DedupScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/dedup.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.DedupSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.dedup.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.DBSNP\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.DBSNPIdx\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf.idx\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.HaplotyperEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/HaplotyperEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.HaplotyperExtraOptionsString\n",
      "value:\t\"'--emit_mode variant --gq_bands 1-60,60-99/19,99 --min_base_qual 10 --pcr_indel_model CONSERVATIVE --phasing 1 --ploidy 2 --prune_factor 2'\"\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.HaplotyperHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.HaplotyperScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/haplotyper.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.HaplotyperSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.HaplotyperVCFSourceField\n",
      "value:\t/projects/bioinformatics/DEL/Config/HaplotyperVCFSourceField.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.Ref\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefAmb\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.amb\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefAnn\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.ann\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefBwt\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.bwt\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefDict\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.dict\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefFai\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.fai\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefPac\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.pac\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.RefSa\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.sa\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.haplotype.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.MergeBamEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/MergeBamEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.MergeBamScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/merge_bams.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.MergeHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.MergeSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.merge.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RealignEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/RealignEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RealignHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RealignSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RealignmentKnownSites\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RealignmentScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/realignment.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.Ref\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefAmb\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.amb\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefAnn\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.ann\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefBwt\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.bwt\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefDict\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.dict\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefFai\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.fai\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefPac\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.pac\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.RefSa\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.sa\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.realign.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.PairedEnd\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.Adapters\n",
      "value:\t/projects/bioinformatics/DEL/Inputs/TruSeqAdaptors.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdapt\n",
      "value:\t/usr/local/apps/bioapps/python/Python-3.6.1/bin\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.CutAdaptThreads\n",
      "value:\t8\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.PairedEnd\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.TrimEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/TrimEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.TrimHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSeqScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/trim_sequences.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_paired.TrimSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.Adapters\n",
      "value:\t/projects/bioinformatics/DEL/Inputs/TruSeqAdaptors.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.CutAdapt\n",
      "value:\t/usr/local/apps/bioapps/python/Python-3.6.1/bin\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.CutAdaptThreads\n",
      "value:\t8\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.PairedEnd\n",
      "value:\ttrue\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.TrimEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/TrimEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.TrimHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.TrimSeqScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/trim_sequences.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.trimseq.TRIMSEQ_single.TrimSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.AnnotateText\n",
      "value:\t\"'--annotation DP --annotation QD --annotation FS --annotation SOR --annotation MQ --annotation MQRankSum --annotation ReadPosRankSum'\"\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.BashPreamble\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shell_preamble.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.BashSharedFunctions\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/shared_functions.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.DebugMode\n",
      "value:\t\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.Ref\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefAmb\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.amb\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefAnn\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.ann\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefBwt\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.bwt\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefDict\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.dict\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefFai\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.fai\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefPac\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.pac\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.RefSa\n",
      "value:\t/projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Homo_sapiens_assembly38.fasta.sa\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.SampleName\n",
      "value:\tWGS_chr1_5X_E0.005_chunk1\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.Sentieon\n",
      "value:\t/usr/local/apps/bioapps/sentieon/sentieon-genomics-201808.01\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.SentieonThreads\n",
      "value:\t40\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.VqsrEnvProfile\n",
      "value:\t/projects/bioinformatics/DEL/Config/VqsrEnvProfile.file\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.VqsrHardMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.VqsrIndelResourceString\n",
      "value:\t\"'--resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf --resource_param dbsnp,known=true,training=false,truth=false,prior=2.0 --resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/Mills_and_1000G_gold_standard.indels.hg38.vcf --resource_param Mills,known=false,training=true,truth=true,prior=12.0'\"\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.VqsrScript\n",
      "value:\t/projects/bioinformatics/DEL/MayomicsVC/src/shell/vqsr.sh\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.VqsrSnpResourceString\n",
      "value:\t\"'--resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/1000G_phase1.snps.high_confidence.hg38.vcf.gz --resource_param 1000G,known=false,training=true,truth=false,prior=10.0 --resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/1000G_omni2.5.hg38.vcf.gz --resource_param omni,known=false,training=true,truth=false,prior=12.0 --resource /projects/mgc/Project_1/MayomicsVC_MayoTesting/Reference/dbsnp_138.hg38.vcf --resource_param dbsnp,known=true,training=false,truth=false,prior=2.0 --resource /projects/bioinformatics/DataPacks/human/gatk_bundle_Oct_2017/gatk_bundle_hg38/hapmap_3.3.hg38.vcf.gz --resource_param hapmap,known=false,training=true,truth=true,prior=15.0'\"\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n",
      " key :\tGermlineMasterWF.vqsr.VqsrSoftMemLimit\n",
      "value:\t100G\n",
      "<class 'str'>\t<class 'str'>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "full_path_json = '../data/Jsons/GermlineMasterWorkflow.FilledIn.json'\n",
    "if os.path.isfile(full_path_json):\n",
    "    fp_json_dict = read_json_file(full_path_json)\n",
    "for k, v in fp_json_dict.items():\n",
    "    print(' key :\\t%s\\nvalue:\\t%s\\n%s\\t%s\\n'%(k,v,type(k),type(v)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
